{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "a2jji3LQ0LAp"
   },
   "source": [
    "# **Milestone 3:**\n",
    "Semantic Search with ML and BERT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "BIfNBiQ4cJQ-"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in /home/d4niel92/miniconda3/lib/python3.9/site-packages (4.15.0)\n",
      "Requirement already satisfied: filelock in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (3.4.2)\n",
      "Requirement already satisfied: sacremoses in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (0.0.47)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (2021.11.10)\n",
      "Requirement already satisfied: requests in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (2.27.1)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (6.0)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.1.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (0.4.0)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (1.21.2)\n",
      "Requirement already satisfied: tokenizers<0.11,>=0.10.1 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (0.10.3)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (21.3)\n",
      "Requirement already satisfied: tqdm>=4.27 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers) (4.62.3)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from huggingface-hub<1.0,>=0.1.0->transformers) (4.0.1)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from packaging>=20.0->transformers) (3.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers) (3.3)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers) (2.0.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers) (1.26.7)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers) (2021.10.8)\n",
      "Requirement already satisfied: six in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sacremoses->transformers) (1.16.0)\n",
      "Requirement already satisfied: click in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sacremoses->transformers) (8.0.3)\n",
      "Requirement already satisfied: joblib in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sacremoses->transformers) (1.1.0)\n",
      "Requirement already satisfied: sentence-transformers in /home/d4niel92/miniconda3/lib/python3.9/site-packages (2.1.0)\n",
      "Requirement already satisfied: sentencepiece in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (0.1.96)\n",
      "Requirement already satisfied: scikit-learn in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (1.0.2)\n",
      "Requirement already satisfied: huggingface-hub in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (0.4.0)\n",
      "Requirement already satisfied: torchvision in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (0.11.2)\n",
      "Requirement already satisfied: transformers<5.0.0,>=4.6.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (4.15.0)\n",
      "Requirement already satisfied: tokenizers>=0.10.3 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (0.10.3)\n",
      "Requirement already satisfied: tqdm in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (4.62.3)\n",
      "Requirement already satisfied: numpy in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (1.21.2)\n",
      "Requirement already satisfied: torch>=1.6.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (1.10.1)\n",
      "Requirement already satisfied: nltk in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (3.6.7)\n",
      "Requirement already satisfied: scipy in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sentence-transformers) (1.7.3)\n",
      "Requirement already satisfied: typing-extensions in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from torch>=1.6.0->sentence-transformers) (4.0.1)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers) (6.0)\n",
      "Requirement already satisfied: filelock in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers) (3.4.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers) (21.3)\n",
      "Requirement already satisfied: sacremoses in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers) (0.0.47)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers) (2021.11.10)\n",
      "Requirement already satisfied: requests in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers) (2.27.1)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from packaging>=20.0->transformers<5.0.0,>=4.6.0->sentence-transformers) (3.0.4)\n",
      "Requirement already satisfied: click in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from nltk->sentence-transformers) (8.0.3)\n",
      "Requirement already satisfied: joblib in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from nltk->sentence-transformers) (1.1.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers<5.0.0,>=4.6.0->sentence-transformers) (3.3)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers<5.0.0,>=4.6.0->sentence-transformers) (1.26.7)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers<5.0.0,>=4.6.0->sentence-transformers) (2021.10.8)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from requests->transformers<5.0.0,>=4.6.0->sentence-transformers) (2.0.4)\n",
      "Requirement already satisfied: six in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from sacremoses->transformers<5.0.0,>=4.6.0->sentence-transformers) (1.16.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from scikit-learn->sentence-transformers) (2.2.0)\n",
      "Requirement already satisfied: pillow!=8.3.0,>=5.3.0 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from torchvision->sentence-transformers) (8.4.0)\n",
      "Requirement already satisfied: elasticsearch in /home/d4niel92/miniconda3/lib/python3.9/site-packages (7.16.2)\n",
      "Requirement already satisfied: urllib3<2,>=1.21.1 in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from elasticsearch) (1.26.7)\n",
      "Requirement already satisfied: certifi in /home/d4niel92/miniconda3/lib/python3.9/site-packages (from elasticsearch) (2021.10.8)\n"
     ]
    }
   ],
   "source": [
    "!pip install transformers\n",
    "!pip install -U sentence-transformers\n",
    "!pip install elasticsearch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OCZ64HgIoKrE"
   },
   "source": [
    "### **Importing the require modules**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "KF2VkBMxzqmr"
   },
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import json\n",
    "import torch\n",
    "import numpy as np\n",
    "from elasticsearch import Elasticsearch, helpers\n",
    "from transformers import AutoModel, AutoTokenizer, pipeline\n",
    "from sentence_transformers import SentenceTransformer, util\n",
    "from pprint import pprint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Ou6WJYn28B_B"
   },
   "source": [
    "### **Getting the data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "dVzHTT4Y8YBx"
   },
   "outputs": [],
   "source": [
    "DATA_DIR = '/mnt/c/Users/Daniel/elasticsearch/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "Jlya41XkXNSj",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# load the json file\n",
    "with open(DATA_DIR + '03_data.json', 'r') as outfile:\n",
    "    data = json.load(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "ge-pmp01t_ax",
    "outputId": "0e3e9bbb-cc84-46fe-c63a-2ca2386b21dc"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'section_title': 'Summary',\n",
       "  'text': 'Pandemic,Summary\\nA pandemic (from Greek πᾶν, pan, \"all\" and δῆμος, demos, \"people\") is an epidemic of an infectious disease that has spread across a large region, for instance multiple continents or worldwide, affecting a substantial number of people. A widespread endemic disease with a stable number of infected people is not a pandemic. Widespread endemic diseases with a stable number of infected people such as recurrences of seasonal influenza are generally excluded as they occur simultaneously in large regions of the globe rather than being spread worldwide.\\nThroughout human history, there have been a number of pandemics of diseases such as smallpox and tuberculosis. The most fatal pandemic in recorded history was the Black Death (also known as The Plague), which killed an estimated 75–200 million people in the 14th century. The term was not used yet but was for later pandemics including the 1918 influenza pandemic (Spanish flu). Current pandemics include COVID-19 (SARS-CoV-2) and HIV/AIDS.',\n",
       "  'main_section': 'Summary',\n",
       "  'article_title': 'Pandemic',\n",
       "  'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "  'page_id': 24255,\n",
       "  'tags': 'Pandemic,Summary',\n",
       "  'section_number': 0},\n",
       " {'section_title': 'Definition',\n",
       "  'text': 'Pandemic,Definition\\nA pandemic is an epidemic occurring on a scale that crosses international boundaries, usually affecting people on a worldwide scale. A disease or condition is not a pandemic merely because it is widespread or kills many people; it must also be infectious. For instance, cancer is responsible for many deaths but is not considered a pandemic because the disease is neither infectious nor contagious.',\n",
       "  'main_section': 'Definition',\n",
       "  'article_title': 'Pandemic',\n",
       "  'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "  'page_id': 24255,\n",
       "  'tags': 'Pandemic,Definition',\n",
       "  'section_number': 1}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print sample sentences\n",
    "g = (d for d in data)\n",
    "[next(g) for i in range(2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "hop1XvWc1mow",
    "outputId": "217dda26-bcdd-4620-850f-d43cccab6cf4"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "401\n"
     ]
    }
   ],
   "source": [
    "print(len(data))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IF-hGN80XNSk"
   },
   "source": [
    "### **Create Elasticsearch index**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "IOFVl5wPtadG",
    "outputId": "ca0cf2c6-35d0-404c-ada7-032c7c42761d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\r\n",
      "  \"name\" : \"bf03c9cd752a\",\r\n",
      "  \"cluster_name\" : \"docker-cluster\",\r\n",
      "  \"cluster_uuid\" : \"_vsaivujSDq8OeX4NUymrw\",\r\n",
      "  \"version\" : {\r\n",
      "    \"number\" : \"7.16.3\",\r\n",
      "    \"build_flavor\" : \"default\",\r\n",
      "    \"build_type\" : \"docker\",\r\n",
      "    \"build_hash\" : \"4e6e4eab2297e949ec994e688dad46290d018022\",\r\n",
      "    \"build_date\" : \"2022-01-06T23:43:02.825887787Z\",\r\n",
      "    \"build_snapshot\" : false,\r\n",
      "    \"lucene_version\" : \"8.10.1\",\r\n",
      "    \"minimum_wire_compatibility_version\" : \"6.8.0\",\r\n",
      "    \"minimum_index_compatibility_version\" : \"6.0.0-beta1\"\r\n",
      "  },\r\n",
      "  \"tagline\" : \"You Know, for Search\"\r\n",
      "}\r\n"
     ]
    }
   ],
   "source": [
    "# check if the elasticsearch container is running\n",
    "!curl http://localhost:9200/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "ZyXcbrvAA2B_"
   },
   "outputs": [],
   "source": [
    "# instantiate elasticsearch client\n",
    "client = Elasticsearch()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "EXuZcfUTA95V"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/d4niel92/miniconda3/lib/python3.9/site-packages/elasticsearch/connection/base.py:200: ElasticsearchWarning: Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone. See https://www.elastic.co/guide/en/elasticsearch/reference/7.16/security-minimal-setup.html to enable security.\n",
      "  warnings.warn(message, category=ElasticsearchWarning)\n"
     ]
    }
   ],
   "source": [
    "# use helpers' bulk API to index list of elasticsearch docs\n",
    "response = helpers.bulk(client, data, index='pandemics')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "Mlpc0LgnBYc6"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/3852563500.py:2: DeprecationWarning: Using positional arguments for APIs is deprecated and will be disabled in 8.0.0. Instead use only keyword arguments for all APIs. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  client.cat.count(\"_all\", params={\"format\": \"json\"})\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'epoch': '1643619693', 'timestamp': '09:01:33', 'count': '401'}]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count documents in all indices\n",
    "client.cat.count(\"_all\", params={\"format\": \"json\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "dTU2SsrYBmLk"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/966010936.py:2: DeprecationWarning: Using positional arguments for APIs is deprecated and will be disabled in 8.0.0. Instead use only keyword arguments for all APIs. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  client.indices.get_alias(\"_all\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'pandemics': {'aliases': {}}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check currently available indices\n",
    "client.indices.get_alias(\"_all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "bdInO9TVByfV"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/4085913714.py:2: DeprecationWarning: Using positional arguments for APIs is deprecated and will be disabled in 8.0.0. Instead use only keyword arguments for all APIs. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  client.cat.count(['pandemics'], params={\"format\": \"json\"})\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'epoch': '1643619693', 'timestamp': '09:01:33', 'count': '401'}]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count documents in specific index\n",
    "client.cat.count(['pandemics'], params={\"format\": \"json\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Different queries**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Match query**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a query body and search a string (fuzzy query)\n",
    "question = \"spanish flu\"\n",
    "match_query_body = {\"query\": {\"match\": {\"text\": question}}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/1006523548.py:2: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  docs = client.search(body = match_query_body, index=\"pandemics\", size=15)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'took': 16,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 45, 'relation': 'eq'},\n",
       "  'max_score': 9.171614,\n",
       "  'hits': [{'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': '_b9dr34BCwOBK0LKaHxc',\n",
       "    '_score': 9.171614,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Influenza',\n",
       "     'text': 'Pandemic,Notable outbreaks,Major outbreaks in countries,Influenza\\nThe Greek physician Hippocrates, the \"Father of Medicine\", first described influenza in 412 BC.\\nThe first influenza pandemic to be pathologically described occurred in 1510. Since the pandemic of 1580, influenza pandemics have occurred every 10 to 30 years.\\nThe 1889–1890 flu pandemic, also known as Russian Flu or Asiatic Flu, was first reported in May 1889 in Bukhara, Uzbekistan. By October, it had reached Tomsk and the Caucasus. It rapidly spread west and hit North America in December 1889, South America in February–April 1890, India in February–March 1890, and Australia in March–April 1890. The H3N8 and H2N2 subtypes of the Influenza A virus have each been identified as possible causes. It had a very high attack and mortality rate, causing around a million fatalities.\\nThe \"Spanish flu\", 1918–1919. First identified early in March 1918 in U.S. troops training at Camp Funston, Kansas. By October 1918, it had spread to become a worldwide pandemic on all continents, and eventually infected about one-third of the world\\'s population (or ≈500 million persons). Unusually deadly and virulent, it ended almost as quickly as it began, vanishing completely within 18 months. Within six months, some 50 million people were dead; some estimates put the total number of fatalities worldwide at over twice that number. About 17 million died in India, 675,000 in the United States, and 200,000 in the United Kingdom. The virus that caused Spanish flu was also implicated as a cause of encephalitis lethargica in children. The virus was recently reconstructed by scientists at the CDC studying remains preserved by the Alaskan permafrost. The H1N1 virus has a small but crucial structure that is similar to the Spanish flu.\\nThe \"Asian Flu\", 1957–58. A H2N2 virus first identified in China in late February 1957. It caused about two million deaths globally.\\nThe \"Hong Kong Flu\", 1968–69. A H3N2 virus first detected in Hong Kong in early 1968 and spread across the world, lasting until 1972. This pandemic killed approximately one million people worldwide.\\nThe \"Swine Flu\", 2009–10. An H1N1 virus first detected in Mexico in early 2009. Estimates for the mortality of this pandemic range from 150 to 500 thousand.',\n",
       "     'main_section': 'Notable outbreaks',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Notable outbreaks,Major outbreaks in countries,Influenza',\n",
       "     'section_number': 12}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'or9dr34BCwOBK0LKaH1i',\n",
       "    '_score': 8.41239,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'In popular culture',\n",
       "     'text': 'Disease X,In popular culture,Candidates\\nIn 2018, the Museum of London ran an exhibition titled \"Disease X: London\\'s next epidemic?\", hosted for the centenary of the Spanish flu epidemic from 1918.The term features in the title of several fiction books that involve global pandemic diseases, such as Disease  (2020), and Disease X: The Outbreak (2019).',\n",
       "     'main_section': 'In popular culture',\n",
       "     'article_title': 'Disease X',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Disease_X',\n",
       "     'page_id': 56809815,\n",
       "     'tags': 'Disease X,In popular culture,Candidates',\n",
       "     'section_number': 9}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Ir9dr34BCwOBK0LKaH5n',\n",
       "    '_score': 7.5050826,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Summary',\n",
       "     'text': 'Swine influenza,Summary\\nSwine influenza is an infection caused by any one of several types of swine influenza viruses. Swine influenza virus (SIV) or swine-origin influenza virus (S-OIV) is any strain of the influenza family of viruses that is endemic in pigs. As of 2009, the known SIV strains include influenza C and the subtypes of  influenza A known as H1N1, H1N2, H2N1, H3N1, H3N2, and H2N3.\\nSwine influenza virus is common throughout pig populations worldwide. Transmission of the virus from pigs to humans is not common and does not always lead to human flu, often resulting only in the production of antibodies in the blood. If transmission does cause human flu, it is called zoonotic swine flu. People with regular exposure to pigs are at increased risk of swine flu infection.\\nAround the mid-20th century, identification of influenza subtypes became possible, allowing accurate diagnosis of transmission to humans. Since then, only 50 such transmissions have been confirmed. These strains of swine flu rarely pass from human to human. Symptoms of zoonotic swine flu in humans are similar to those of influenza and of influenza-like illness in general, namely chills, fever, sore throat, muscle pains, severe headache, coughing, weakness, shortness of breath, and general discomfort.\\nIt is estimated that in the 2009 flu pandemic 11–21% of the then global population (of about 6.8 billion), or around 700 million to 1.4 billion people, contracted the illness—more in absolute terms than the Spanish flu pandemic. There were 18,449 confirmed fatalities. However, in a 2012 study, the CDC estimated more than 284,000 possible fatalities worldwide, with range from 150,000 to 575,000.\\nIn August 2010, the World Health Organization declared the swine flu pandemic officially over.Subsequent cases of swine flu were reported in India in 2015, with over 31,156 positive test cases and 1,841 deaths up to March 2015.',\n",
       "     'main_section': 'Summary',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,Summary',\n",
       "     'section_number': 0}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Cr9dr34BCwOBK0LKaH1c',\n",
       "    '_score': 7.022901,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'H5N1 (Avian flu)',\n",
       "     'text': 'Pandemic,Concerns about future pandemics,Concerning diseases,H5N1 (Avian flu)\\nIn February 2004, avian influenza virus was detected in birds in Vietnam, increasing fears of the emergence of new variant strains. It is feared that if the avian influenza virus combines with a human influenza virus (in a bird or a human), the new subtype created could be both highly contagious and highly lethal in humans. Such a subtype could cause a global influenza pandemic, similar to the Spanish flu or the lower mortality pandemics such as the Asian Flu and the Hong Kong Flu.\\nFrom October 2004 to February 2005, some 3,700 test kits of the 1957 Asian Flu virus were accidentally spread around the world from a lab in the U.S.In May 2005, scientists urgently called upon nations to prepare for a global influenza pandemic that could strike as much as 20% of the world\\'s population.In October 2005, cases of the avian flu (the deadly strain H5N1) were identified in Turkey. EU Health Commissioner Markos Kyprianou said: \"We have received now confirmation that the virus found in Turkey is an avian flu H5N1 virus. There is a direct relationship with viruses found in Russia, Mongolia and China.\" Cases of bird flu were also identified shortly thereafter in Romania, and then Greece. Possible cases of the virus have also been found in Croatia, Bulgaria and the United Kingdom.By November 2007, numerous confirmed cases of the H5N1 strain had been identified across Europe. However, by the end of October, only 59 people had died as a result of H5N1, which was atypical of previous influenza pandemics.\\nAvian flu cannot be categorized as a \"pandemic\" because the virus cannot yet cause sustained and efficient human-to-human transmission. Cases so far are recognized to have been transmitted from bird to human, but as of December 2006 there had been few (if any) cases of proven human-to-human transmission. Regular influenza viruses establish infection by attaching to receptors in the throat and lungs, but the avian influenza virus can attach only to receptors located deep in the lungs of humans, requiring close, prolonged contact with infected patients, and thus limiting person-to-person transmission.',\n",
       "     'main_section': 'Concerns about future pandemics',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Concerns about future pandemics,Concerning diseases,H5N1 (Avian flu)',\n",
       "     'section_number': 30}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': '9L9dr34BCwOBK0LKaHxb',\n",
       "    '_score': 6.6645823,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Summary',\n",
       "     'text': 'Pandemic,Summary\\nA pandemic (from Greek πᾶν, pan, \"all\" and δῆμος, demos, \"people\") is an epidemic of an infectious disease that has spread across a large region, for instance multiple continents or worldwide, affecting a substantial number of people. A widespread endemic disease with a stable number of infected people is not a pandemic. Widespread endemic diseases with a stable number of infected people such as recurrences of seasonal influenza are generally excluded as they occur simultaneously in large regions of the globe rather than being spread worldwide.\\nThroughout human history, there have been a number of pandemics of diseases such as smallpox and tuberculosis. The most fatal pandemic in recorded history was the Black Death (also known as The Plague), which killed an estimated 75–200 million people in the 14th century. The term was not used yet but was for later pandemics including the 1918 influenza pandemic (Spanish flu). Current pandemics include COVID-19 (SARS-CoV-2) and HIV/AIDS.',\n",
       "     'main_section': 'Summary',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Summary',\n",
       "     'section_number': 0}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': '8b9dr34BCwOBK0LKaH1k',\n",
       "    '_score': 6.203658,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Guidelines',\n",
       "     'text': 'Pandemic severity index,Guidelines\\nThe index focuses less on how likely a disease will spread worldwide – that is, become a pandemic – and more upon how severe the epidemic actually is.\\nThe main criterion used to measure pandemic severity will be case-fatality rate (CFR), the percentage of deaths out of the total reported cases of the disease.The actual implementation of PSI alerts was expected to occur after the World Health Organization (WHO) announces phase 6 influenza transmission (human to human) in the United States. This would probably result in immediate announcement of a PSI level 3–4 situation.The analogy of \"category\" levels were introduced to provide an understandable connection to hurricane classification schemes, with specific reference to the recent aftermath of Hurricane Katrina.\\nLike the Saffir–Simpson Hurricane Scale, the PSI ranges from 1 to 5, with Category 1 pandemics being most mild (equivalent to seasonal flu) and level 5 being reserved for the most severe \"worst-case\" scenario pandemics (such as the 1918 Spanish flu).\\n\\nThe report recommends four primary social distancing measures for slowing down a pandemic:\\n\\nIsolation and treatment of people who have suspected or confirmed cases of pandemic influenza\\nVoluntary home quarantine of household contacts of those with suspected or confirmed pandemic influenza\\nDismissing school classes and closing daycare centers\\nChanging work schedules and canceling large public gatheringsThese actions, when implemented, can have an overall effect of reducing the number of new cases of the disease; but they can carry potentially adverse consequences in terms of community and social disruption. The measures should have the most noticeable impact if implemented uniformly by organizations and governments across the US.',\n",
       "     'main_section': 'Guidelines',\n",
       "     'article_title': 'Pandemic severity index',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic_severity_index',\n",
       "     'page_id': 9291245,\n",
       "     'tags': 'Pandemic severity index,Guidelines',\n",
       "     'section_number': 3}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Cr9dr34BCwOBK0LKaH5l',\n",
       "    '_score': 5.680986,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Later',\n",
       "     'text': '1929–1930 psittacosis pandemic,North America,Africa,Later\\n43 of the 74 foci in the U.S. were traced to contact with Amazon parrots. Links were traced to Japan, Caribbean, Germany, Central America and South America. Between November 1929 and May 1930, the U.S. recorded 169 cases, of which 33 died. New York was the centre of the East Coast bird trade. However, the principal ports of entry for Australian budgerigars was San Francisco and Los Angeles. Later, it was discovered that the main source was domestic lovebirds raised in hundreds of independent Californian aviaries by breeders who were supplementing their incomes following the recent Wall Street Crash. The winter of 1929 was also witnessing an influenza epidemic and there were fears of a recurrence of Spanish flu,which added to the depressive effects. In this context, peddlers travelled door-to-door with “lovebirds” for housewives and widows. As a result, most victims in the U.S. were women. The realisation and connections of the various outbreaks may not have become apparent had it not been for the press. Likewise, the \"hysteria\" and heightened public concern surrounding the pandemic may not have occurred had it not been for headlines such as “Killed By A Pet Parrot.”The establishment of the National Institutes of Health is directly linked to the outbreak that occurred in Maryland. Its story was retold in Paul de Kruif\\'s, Men Against Death (1933).',\n",
       "     'main_section': 'North America',\n",
       "     'article_title': '1929–1930 psittacosis pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/1929%E2%80%931930_psittacosis_pandemic',\n",
       "     'page_id': 60428423,\n",
       "     'tags': '1929–1930 psittacosis pandemic,North America,Africa,Later',\n",
       "     'section_number': 12}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': '_r9dr34BCwOBK0LKaHxc',\n",
       "    '_score': 4.6167145,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Typhus',\n",
       "     'text': 'Pandemic,Notable outbreaks,Major outbreaks in countries,Typhus\\nTyphus is sometimes called \"camp fever\" because of its pattern of flaring up in times of strife. (It is also known as \"gaol fever\" and \"ship fever\", for its habits of spreading wildly in cramped quarters, such as jails and ships.) Emerging during the Crusades, it had its first impact in Europe in 1489, in Spain. During fighting between the Christian Spaniards and the Muslims in Granada, the Spanish lost 3,000 to war casualties, and 20,000 to typhus. In 1528, the French lost 18,000 troops in Italy, and lost supremacy in Italy to the Spanish. In 1542, 30,000 soldiers died of typhus while fighting the Ottomans in the Balkans.\\nDuring the Thirty Years\\' War (1618–1648), about eight million Germans were killed by bubonic plague and typhus. The disease also played a major role in the destruction of Napoleon\\'s Grande Armée in Russia in 1812. During the retreat from Moscow, more French military personnel died of typhus than were killed by the Russians. Of the 450,000 soldiers who crossed the Neman on 25 June 1812, fewer than 40,000 returned. More military personnel were killed from 1500 to 1914 by typhus than from military action. In early 1813, Napoleon raised a new army of 500,000 to replace his Russian losses. In the campaign of that year, more than 219,000 of Napoleon\\'s soldiers died of typhus. Typhus played a major factor in the Great Famine of Ireland. During World War I, typhus epidemics killed more than 150,000 in Serbia. There were about 25 million infections and 3 million deaths from epidemic typhus in Russia from 1918 to 1922. Typhus also killed numerous prisoners in the Nazi concentration camps and Soviet prisoner of war camps during World War II. More than 3.5 million Soviet POWs died out of the 5.7 million in Nazi custody.',\n",
       "     'main_section': 'Notable outbreaks',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Notable outbreaks,Major outbreaks in countries,Typhus',\n",
       "     'section_number': 13}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Qb9dr34BCwOBK0LKaH5o',\n",
       "    '_score': 4.388603,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'External links',\n",
       "     'text': 'Swine influenza,External links,Further reading\\nOfficial swine flu advice and latest information from the UK National Health Service\\n8 minute video answering common questions about the subject on fora.tv\\nSwine flu charts and maps Numeric analysis and approximation of current active cases\\n\"Swine Influenza\" disease card on World Organisation for Animal Health\\nWorried about swine flu? Then you should be terrified about the regular flu.\\nCenters for Disease Control and Prevention (CDC) – Swine Flu\\nCenter for Infectious Disease Research and Policy – Novel H1N1 influenza resource list\\nPandemic Flu US Government Site\\nWorld Health Organization (WHO): Swine influenza\\nMedical Encyclopedia Medline Plus: Swine Flu\\nHealth-EU portal EU response to influenza\\nEuropean Commission – Public Health EU coordination on Pandemic (H1N1) 2009',\n",
       "     'main_section': 'External links',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,External links,Further reading',\n",
       "     'section_number': 38}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'JL9dr34BCwOBK0LKaH5n',\n",
       "    '_score': 4.179197,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Humans',\n",
       "     'text': 'Swine influenza,Signs and symptoms,Humans\\nDirect transmission of a swine flu virus from pigs to humans is occasionally possible (zoonotic swine flu). In all, 50 cases are known to have occurred since the first report in medical literature in 1958, which have resulted in a total of six deaths. Of these six people, one was pregnant, one had leukemia, one had Hodgkin\\'s lymphoma and two were known to be previously healthy. One of these had unknown whereabouts. Despite these apparently low numbers of infections, the true rate of infection may be higher, since most cases only cause a very mild disease, and will probably never be reported or diagnosed.\\n\\nAccording to the United States Centers for Disease Control and Prevention (CDC), in humans the symptoms of the 2009 \"swine flu\" H1N1 virus are similar to influenza and influenza-like illness in general. Symptoms include fever, cough, sore throat, watery eyes, body aches, shortness of breath, headache, weight loss, chills, sneezing, runny nose, coughing, dizziness, abdominal pain, lack of appetite, and fatigue. The 2009 outbreak showed an increased percentage of patients reporting diarrhea and vomiting as well. The 2009 H1N1 virus is not zoonotic swine flu, as it is not transmitted from pigs to humans, but from person to person through airborne droplets.Because these symptoms are not specific to swine flu, a differential diagnosis of probable swine flu requires not only symptoms, but also a high likelihood of swine flu due to the person\\'s recent and past medical history. For example, during the 2009 swine flu outbreak in the United States, the CDC advised physicians to \"consider swine influenza infection in the differential diagnosis of patients with acute febrile respiratory illness who have either been in contact with persons with confirmed swine flu, or who were in one of the five U.S. states that have reported swine flu cases or in Mexico during the seven days preceding their illness onset.\" A diagnosis of confirmed swine flu requires laboratory testing of a respiratory sample (a simple nose and throat swab).The most common cause of death is respiratory failure. Other causes of death are pneumonia (leading to sepsis), high fever (leading to neurological problems), dehydration (from excessive vomiting and diarrhea), electrolyte imbalance and kidney failure. Fatalities are more likely in young children and the elderly.',\n",
       "     'main_section': 'Signs and symptoms',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,Signs and symptoms,Humans',\n",
       "     'section_number': 2}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'O79dr34BCwOBK0LKaH5o',\n",
       "    '_score': 4.0179915,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': '2015 and 2017 India',\n",
       "     'text': 'Swine influenza,History,Outbreaks,2015 and 2017 India\\nSwine flu outbreaks were reported in India in late 2014 and early 2015. As of March 19, 2015 the disease has affected 31,151 people and claimed over 1,841 lives. The largest number of reported cases and deaths due to the disease occurred in the western part of India including states like Delhi, Madhya Pradesh, Rajasthan, and Gujarat.\\nResearchers of MIT have claimed that the swine flu has mutated in India to a more virulent version with changes in Hemagglutinin protein, contradicting earlier research by Indian researchers.There was another outbreak in India in 2017. The states of Maharashtra and Gujarat were the worst affected. Gujarat high court has given Gujarat government instructions to control deaths by swine flu. 1,090 people died of swine flu in India in 2017 until August 26, 2017.',\n",
       "     'main_section': 'History',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,History,Outbreaks,2015 and 2017 India',\n",
       "     'section_number': 30}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Jb9dr34BCwOBK0LKaH1d',\n",
       "    '_score': 4.000816,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Military concerns',\n",
       "     'text': \"Antonine Plague,Military concerns,Impact\\nSome direct effects of the contagion stand out. When imperial forces moved east, under the command of Emperor Verus, after the forces of Vologases IV of Parthia attacked Armenia, the Romans' defense of the eastern territories was hampered when large numbers of troops succumbed to the disease. According to the 5th-century Spanish writer Paulus Orosius, many towns and villages in the Italian Peninsula and the European provinces lost all of their inhabitants. As the disease swept north to the Rhine, it also infected Germanic and Gallic peoples outside the empire's borders. For years, those northern groups had pressed south in search of more lands to sustain their growing populations. With their ranks thinned by the epidemic, Roman armies were now unable to push the tribes back. From 167 to his death, Marcus Aurelius personally commanded legions near the Danube, trying, with only partial success, to control the advance of Germanic peoples across the river. A major offensive against the Marcomanni was postponed to 169 because of a shortage of imperial troops.\",\n",
       "     'main_section': 'Military concerns',\n",
       "     'article_title': 'Antonine Plague',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Antonine_Plague',\n",
       "     'page_id': 845862,\n",
       "     'tags': 'Antonine Plague,Military concerns,Impact',\n",
       "     'section_number': 4}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Ob9dr34BCwOBK0LKaH5o',\n",
       "    '_score': 4.000139,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': '2007 Philippines',\n",
       "     'text': 'Swine influenza,History,Outbreaks,2007 Philippines\\nOn August 20, 2007, Department of Agriculture officers investigated the outbreak of swine flu in Nueva Ecija and central Luzon, Philippines. The mortality rate is less than 10% for swine flu, unless there are complications like hog cholera. On July 27, 2007, the Philippine National Meat Inspection Service (NMIS) raised a hog cholera \"red alert\" warning over Metro Manila and five regions of Luzon after the disease spread to backyard pig farms in Bulacan and Pampanga, even if they tested negative for the swine flu virus.',\n",
       "     'main_section': 'History',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,History,Outbreaks,2007 Philippines',\n",
       "     'section_number': 28}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Pb9dr34BCwOBK0LKaH5o',\n",
       "    '_score': 3.978426,\n",
       "    '_source': {'section_title': '2016 Pakistan',\n",
       "     'text': 'Swine influenza,History,Outbreaks,2016 Pakistan\\nSeven cases of swine flu were reported in Punjab province of Pakistan, mainly in the city of Multan, in January 2017. Cases of swine flu were also reported in Lahore and Faisalabad.',\n",
       "     'main_section': 'History',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,History,Outbreaks,2016 Pakistan',\n",
       "     'section_number': 32}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'QL9dr34BCwOBK0LKaH5o',\n",
       "    '_score': 3.9171398,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'H1N1 virus pandemic history',\n",
       "     'text': 'Swine influenza,History,Outbreaks,H1N1 virus pandemic history\\nA 2008 study discussed the evolutionary origin of the flu strain of swine origin (S-OIV).\\nAccording to this study, the phylogenetic origin of the flu virus that caused the 2009 pandemics can be traced before 1918. Around 1918, the ancestral virus, of avian origin, crossed the species boundaries and infected humans as human H1N1. The same phenomenon took place soon after in America, where the human virus infected pigs; it led to the emergence of the H1N1 swine strain, which later became known as swine flu. Genetic coding of H1N1 shows it is a combination of segments of four influenza viruses forming a novel strain:\\n\\nNorth American Swine (30.6%) – pig origin\\nNorth American Avian (34.4%) – bird origin\\nHuman influenza strain (17.5%)\\nEuroasian swine (17.5%) – Pig originQuadruple genetic re-assortment  – coinfection with influenza viruses from diverse animal species.\\nDue to coinfection, the viruses are able to interact, mutate, and form a new strain to which host has variable immunity.\\nNew events of reassortment were not reported until 1968, when the avian strain H1N1 infected humans again; this time the virus met the strain H2N2, and the reassortment originated the strain H3N2. This strain has remained as a stable flu strain until now.\\nThe mid-1970s were important for the evolution of flu strains. First, the re-emergence of the human H1N1 strain became a seasonal strain. Then, a small outbreak of swine H1N1 occurred in humans, and finally, the human H2N2 strain apparently became extinct. Around 1979, the avian H1N1 strain infected pigs and gave rise to Euroasiatic swine flu and H1N1 Euroasiatic swine virus, which is still being transmitted in swine populations.\\nThe critical moment for the 2009 outbreak was between 1990 and 1993. A triple reassortment event in a pig host of North American H1N1 swine virus, the human H3N2 virus and avian H1N1 virus generated the swine H1N2 strain. In 2009, when the virus H1N2 co-infected a human host at the same time as the Euroasiatic H1N1 swine strain a new human H1N1 strain emerged, which caused the 2009 pandemic.\\nSwine flu spread very rapidly worldwide due to its high human-to-human transmission rate and due to the frequency of air travel.',\n",
       "     'main_section': 'History',\n",
       "     'article_title': 'Swine influenza',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Swine_influenza',\n",
       "     'page_id': 1516915,\n",
       "     'tags': 'Swine influenza,History,Outbreaks,H1N1 virus pandemic history',\n",
       "     'section_number': 35}}]}}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# submit the search query to elasticsearch\n",
    "docs = client.search(body = match_query_body, index=\"pandemics\", size=15)\n",
    "\n",
    "print(type(docs))\n",
    "docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print the three top results\n",
    "def top_three_results(docs, responses):\n",
    "    for doc in docs[\"hits\"][\"hits\"][:3]:\n",
    "        article_title = doc['_source']['article_title']\n",
    "        section_title = doc['_source']['section_title']\n",
    "        score = doc['_score']\n",
    "        idx = doc['_index']\n",
    "        text = doc['_source']['text']\n",
    "        url = doc['_source']['source_url']\n",
    "\n",
    "        responses.append({'article_title': article_title,\n",
    "                         'section_title': section_title,\n",
    "                         'text': text,\n",
    "                          'url': url,\n",
    "                          'score': score,\n",
    "                         })\n",
    "    return responses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'article_title': 'Pandemic',\n",
      "  'score': 9.171614,\n",
      "  'section_title': 'Influenza',\n",
      "  'text': 'Pandemic,Notable outbreaks,Major outbreaks in countries,Influenza\\n'\n",
      "          'The Greek physician Hippocrates, the \"Father of Medicine\", first '\n",
      "          'described influenza in 412 BC.\\n'\n",
      "          'The first influenza pandemic to be pathologically described '\n",
      "          'occurred in 1510. Since the pandemic of 1580, influenza pandemics '\n",
      "          'have occurred every 10 to 30 years.\\n'\n",
      "          'The 1889–1890 flu pandemic, also known as Russian Flu or Asiatic '\n",
      "          'Flu, was first reported in May 1889 in Bukhara, Uzbekistan. By '\n",
      "          'October, it had reached Tomsk and the Caucasus. It rapidly spread '\n",
      "          'west and hit North America in December 1889, South America in '\n",
      "          'February–April 1890, India in February–March 1890, and Australia in '\n",
      "          'March–April 1890. The H3N8 and H2N2 subtypes of the Influenza A '\n",
      "          'virus have each been identified as possible causes. It had a very '\n",
      "          'high attack and mortality rate, causing around a million '\n",
      "          'fatalities.\\n'\n",
      "          'The \"Spanish flu\", 1918–1919. First identified early in March 1918 '\n",
      "          'in U.S. troops training at Camp Funston, Kansas. By October 1918, '\n",
      "          'it had spread to become a worldwide pandemic on all continents, and '\n",
      "          \"eventually infected about one-third of the world's population (or \"\n",
      "          '≈500 million persons). Unusually deadly and virulent, it ended '\n",
      "          'almost as quickly as it began, vanishing completely within 18 '\n",
      "          'months. Within six months, some 50 million people were dead; some '\n",
      "          'estimates put the total number of fatalities worldwide at over '\n",
      "          'twice that number. About 17 million died in India, 675,000 in the '\n",
      "          'United States, and 200,000 in the United Kingdom. The virus that '\n",
      "          'caused Spanish flu was also implicated as a cause of encephalitis '\n",
      "          'lethargica in children. The virus was recently reconstructed by '\n",
      "          'scientists at the CDC studying remains preserved by the Alaskan '\n",
      "          'permafrost. The H1N1 virus has a small but crucial structure that '\n",
      "          'is similar to the Spanish flu.\\n'\n",
      "          'The \"Asian Flu\", 1957–58. A H2N2 virus first identified in China in '\n",
      "          'late February 1957. It caused about two million deaths globally.\\n'\n",
      "          'The \"Hong Kong Flu\", 1968–69. A H3N2 virus first detected in Hong '\n",
      "          'Kong in early 1968 and spread across the world, lasting until 1972. '\n",
      "          'This pandemic killed approximately one million people worldwide.\\n'\n",
      "          'The \"Swine Flu\", 2009–10. An H1N1 virus first detected in Mexico in '\n",
      "          'early 2009. Estimates for the mortality of this pandemic range from '\n",
      "          '150 to 500 thousand.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Pandemic'},\n",
      " {'article_title': 'Disease X',\n",
      "  'score': 8.41239,\n",
      "  'section_title': 'In popular culture',\n",
      "  'text': 'Disease X,In popular culture,Candidates\\n'\n",
      "          'In 2018, the Museum of London ran an exhibition titled \"Disease X: '\n",
      "          'London\\'s next epidemic?\", hosted for the centenary of the Spanish '\n",
      "          'flu epidemic from 1918.The term features in the title of several '\n",
      "          'fiction books that involve global pandemic diseases, such as '\n",
      "          'Disease  (2020), and Disease X: The Outbreak (2019).',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Disease_X'},\n",
      " {'article_title': 'Swine influenza',\n",
      "  'score': 7.5050826,\n",
      "  'section_title': 'Summary',\n",
      "  'text': 'Swine influenza,Summary\\n'\n",
      "          'Swine influenza is an infection caused by any one of several types '\n",
      "          'of swine influenza viruses. Swine influenza virus (SIV) or '\n",
      "          'swine-origin influenza virus (S-OIV) is any strain of the influenza '\n",
      "          'family of viruses that is endemic in pigs. As of 2009, the known '\n",
      "          'SIV strains include influenza C and the subtypes of  influenza A '\n",
      "          'known as H1N1, H1N2, H2N1, H3N1, H3N2, and H2N3.\\n'\n",
      "          'Swine influenza virus is common throughout pig populations '\n",
      "          'worldwide. Transmission of the virus from pigs to humans is not '\n",
      "          'common and does not always lead to human flu, often resulting only '\n",
      "          'in the production of antibodies in the blood. If transmission does '\n",
      "          'cause human flu, it is called zoonotic swine flu. People with '\n",
      "          'regular exposure to pigs are at increased risk of swine flu '\n",
      "          'infection.\\n'\n",
      "          'Around the mid-20th century, identification of influenza subtypes '\n",
      "          'became possible, allowing accurate diagnosis of transmission to '\n",
      "          'humans. Since then, only 50 such transmissions have been confirmed. '\n",
      "          'These strains of swine flu rarely pass from human to human. '\n",
      "          'Symptoms of zoonotic swine flu in humans are similar to those of '\n",
      "          'influenza and of influenza-like illness in general, namely chills, '\n",
      "          'fever, sore throat, muscle pains, severe headache, coughing, '\n",
      "          'weakness, shortness of breath, and general discomfort.\\n'\n",
      "          'It is estimated that in the 2009 flu pandemic 11–21% of the then '\n",
      "          'global population (of about 6.8 billion), or around 700 million to '\n",
      "          '1.4 billion people, contracted the illness—more in absolute terms '\n",
      "          'than the Spanish flu pandemic. There were 18,449 confirmed '\n",
      "          'fatalities. However, in a 2012 study, the CDC estimated more than '\n",
      "          '284,000 possible fatalities worldwide, with range from 150,000 to '\n",
      "          '575,000.\\n'\n",
      "          'In August 2010, the World Health Organization declared the swine '\n",
      "          'flu pandemic officially over.Subsequent cases of swine flu were '\n",
      "          'reported in India in 2015, with over 31,156 positive test cases and '\n",
      "          '1,841 deaths up to March 2015.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Swine_influenza'}]\n"
     ]
    }
   ],
   "source": [
    "responses = []\n",
    "pprint(top_three_results(docs=docs, responses=responses))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Term query**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a query body and search for a perfect match\n",
    "## note that you need to use the keyword version of the field if your search query contains more that one word.\n",
    "term_question = \"Zoonotic viruses\"\n",
    "term_query_body = {\"query\": {\"term\": {\"section_title.keyword\": term_question}}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/367153364.py:2: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  term_docs = client.search(body = term_query_body, index=\"pandemics\", size=5)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'took': 10,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 1, 'relation': 'eq'},\n",
       "  'max_score': 5.5909867,\n",
       "  'hits': [{'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'nb9dr34BCwOBK0LKaH1h',\n",
       "    '_score': 5.5909867,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Zoonotic viruses',\n",
       "     'text': 'Disease X,Candidates,Zoonotic viruses\\nOn the addition of Disease X in 2018, the WHO said it could come from many sources citing haemorrhagic fevers and the more recent non-polio enterovirus. However, Røttingen speculated that Disease X would be more likely come from zoonotic transmission (an animal virus that jumps to humans), saying: \"It\\'s a natural process and it is vital that we are aware and prepare. It is probably the greatest risk\". WHO special advisor Professor Marion Koopmans, also noted that the rate at which zoonotic diseases were appearing was accelerating, saying: \"The intensity of animal and human contact is becoming much greater as the world develops. This makes it more likely new diseases will emerge but also modern travel and trade make it much more likely they will spread\".',\n",
       "     'main_section': 'Candidates',\n",
       "     'article_title': 'Disease X',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Disease_X',\n",
       "     'page_id': 56809815,\n",
       "     'tags': 'Disease X,Candidates,Zoonotic viruses',\n",
       "     'section_number': 4}}]}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# submit the search query to elasticsearch\n",
    "term_docs = client.search(body = term_query_body, index=\"pandemics\", size=5)\n",
    "\n",
    "print(type(term_docs))\n",
    "term_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'article_title': 'Disease X',\n",
      "  'score': 5.5909867,\n",
      "  'section_title': 'Zoonotic viruses',\n",
      "  'text': 'Disease X,Candidates,Zoonotic viruses\\n'\n",
      "          'On the addition of Disease X in 2018, the WHO said it could come '\n",
      "          'from many sources citing haemorrhagic fevers and the more recent '\n",
      "          'non-polio enterovirus. However, Røttingen speculated that Disease X '\n",
      "          'would be more likely come from zoonotic transmission (an animal '\n",
      "          'virus that jumps to humans), saying: \"It\\'s a natural process and '\n",
      "          'it is vital that we are aware and prepare. It is probably the '\n",
      "          'greatest risk\". WHO special advisor Professor Marion Koopmans, also '\n",
      "          'noted that the rate at which zoonotic diseases were appearing was '\n",
      "          'accelerating, saying: \"The intensity of animal and human contact is '\n",
      "          'becoming much greater as the world develops. This makes it more '\n",
      "          'likely new diseases will emerge but also modern travel and trade '\n",
      "          'make it much more likely they will spread\".',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Disease_X'}]\n"
     ]
    }
   ],
   "source": [
    "# print the three top results\n",
    "term_responses = []\n",
    "pprint(top_three_results(term_docs, responses=term_responses))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Boolean query**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude sections from the search results\n",
    "bool_question = \"world health organization\"\n",
    "exclude_sections = \"External links\"\n",
    "bool_query_body = {\n",
    "  \"query\": {\n",
    "    \"bool\": {\n",
    "        \"should\": \n",
    "        { \"match\": {\"text\": bool_question}},\n",
    "        \"must_not\": {\n",
    "            \"term\" : { \"section_title.keyword\" : exclude_sections }\n",
    "      },\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/1584265501.py:2: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  bool_docs = client.search(body = bool_query_body, index=\"pandemics\", size=5)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'took': 14,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 158, 'relation': 'eq'},\n",
       "  'max_score': 7.7753215,\n",
       "  'hits': [{'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'pb9dr34BCwOBK0LKaH1i',\n",
       "    '_score': 7.7753215,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Summary',\n",
       "     'text': 'Johns Hopkins Center for Health Security,Summary\\nThe Johns Hopkins Center for Health Security (abbreviated CHS; previously the UPMC Center for Health Security, the Center for Biosecurity of UPMC, and the Johns Hopkins Center for Civilian Biodefense Strategies) is an independent, nonprofit organization of the Johns Hopkins Bloomberg School of Public Health, and part of the Environmental Health and Engineering department. It is concerned with the areas of health consequences from epidemics and disasters as well as averting biological weapons development, and implications of biosecurity for the bioeconomy. It is a think tank that does policy research and gives policy recommendations to the United States government as well as the World Health Organization and the UN Biological Weapons Convention.',\n",
       "     'main_section': 'Summary',\n",
       "     'article_title': 'Johns Hopkins Center for Health Security',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Johns_Hopkins_Center_for_Health_Security',\n",
       "     'page_id': 11008433,\n",
       "     'tags': 'Johns Hopkins Center for Health Security,Summary',\n",
       "     'section_number': 0}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Dr9dr34BCwOBK0LKaH5l',\n",
       "    '_score': 7.602386,\n",
       "    '_source': {'section_title': 'Further reading',\n",
       "     'text': '1929–1930 psittacosis pandemic,Further reading,References\\nAntonin Morange (1895). De la psittacose ou infection spéciale déterminée par les perruches. Médecine de Paris.\\nWeekly Epidemiological Record. World Health Organization; IRIS',\n",
       "     'main_section': 'Further reading',\n",
       "     'article_title': '1929–1930 psittacosis pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/1929%E2%80%931930_psittacosis_pandemic',\n",
       "     'page_id': 60428423,\n",
       "     'tags': '1929–1930 psittacosis pandemic,Further reading,References',\n",
       "     'section_number': 18}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'k79dr34BCwOBK0LKaH1h',\n",
       "    '_score': 7.26242,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Data and graphs',\n",
       "     'text': 'COVID-19 pandemic,External links,Data and graphs\\nCoronavirus disease (COVID-2019) situation reports and map by the World Health Organization\\nCoronavirus Resource Center, map, and historical data by Johns Hopkins University\\nHistorical data about COVID-19 cases published by the European Centre for Disease Prevention and Control (ECDC)\\nWorld Travel Restrictions based on WFP data\\nCoronavirus Observer based on Johns Hopkins University data\\nCOVID-19 coronavirus pandemic published by Worldometer\\nCoronavirus Disease (COVID-19) Statistics and Research published by Our World in Data\\nCOVID-19 Projections for many countries published by Institute for Health Metrics and Evaluation\\nSpreadsheets: Country comparisons\\nSpreadsheets: Historical data with additional computations\\nEpidemic Calculator\\nStat News COVID-19 Tracker',\n",
       "     'main_section': 'External links',\n",
       "     'article_title': 'COVID-19 pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/COVID-19_pandemic',\n",
       "     'page_id': 62750956,\n",
       "     'tags': 'COVID-19 pandemic,External links,Data and graphs',\n",
       "     'section_number': 81}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'kb9dr34BCwOBK0LKaH1h',\n",
       "    '_score': 7.253867,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Health agencies',\n",
       "     'text': 'COVID-19 pandemic,External links,Health agencies\\nCOVID-19 (Questions & Answers, instructional videos; Facts/MythBusters) by the World Health Organization\\nCOVID-19 by the Government of Canada\\nCOVID-19 by the National Health Commission, China\\nCOVID-19 (Q&A) by the European Centre for Disease Prevention and Control\\nCOVID-19 (Q&A) by the Ministry of Health, Singapore\\nCOVID-19 (Q&A) by the U.S. Centers for Disease Control\\nNIH COVID-19 Portfolio by the U.S. National Institutes of Health\\nCoronavirus Disease 2019 (Q&A) by the U.S. National Institute for Occupational Safety and Health',\n",
       "     'main_section': 'External links',\n",
       "     'article_title': 'COVID-19 pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/COVID-19_pandemic',\n",
       "     'page_id': 62750956,\n",
       "     'tags': 'COVID-19 pandemic,External links,Health agencies',\n",
       "     'section_number': 79}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'C79dr34BCwOBK0LKaH1c',\n",
       "    '_score': 6.9860735,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Zika virus',\n",
       "     'text': 'Pandemic,Concerns about future pandemics,Concerning diseases,Zika virus\\nAn outbreak of Zika virus began in 2015 and strongly intensified throughout the start of 2016, with more than 1.5 million cases across more than a dozen countries in the Americas. The World Health Organization warned that Zika had the potential to become an explosive global pandemic if the outbreak was not controlled.',\n",
       "     'main_section': 'Concerns about future pandemics',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Concerns about future pandemics,Concerning diseases,Zika virus',\n",
       "     'section_number': 31}}]}}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# submit the search query to elasticsearch\n",
    "bool_docs = client.search(body = bool_query_body, index=\"pandemics\", size=5)\n",
    "\n",
    "print(type(bool_docs))\n",
    "bool_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'article_title': 'Johns Hopkins Center for Health Security',\n",
      "  'score': 7.7753215,\n",
      "  'section_title': 'Summary',\n",
      "  'text': 'Johns Hopkins Center for Health Security,Summary\\n'\n",
      "          'The Johns Hopkins Center for Health Security (abbreviated CHS; '\n",
      "          'previously the UPMC Center for Health Security, the Center for '\n",
      "          'Biosecurity of UPMC, and the Johns Hopkins Center for Civilian '\n",
      "          'Biodefense Strategies) is an independent, nonprofit organization of '\n",
      "          'the Johns Hopkins Bloomberg School of Public Health, and part of '\n",
      "          'the Environmental Health and Engineering department. It is '\n",
      "          'concerned with the areas of health consequences from epidemics and '\n",
      "          'disasters as well as averting biological weapons development, and '\n",
      "          'implications of biosecurity for the bioeconomy. It is a think tank '\n",
      "          'that does policy research and gives policy recommendations to the '\n",
      "          'United States government as well as the World Health Organization '\n",
      "          'and the UN Biological Weapons Convention.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Johns_Hopkins_Center_for_Health_Security'},\n",
      " {'article_title': '1929–1930 psittacosis pandemic',\n",
      "  'score': 7.602386,\n",
      "  'section_title': 'Further reading',\n",
      "  'text': '1929–1930 psittacosis pandemic,Further reading,References\\n'\n",
      "          'Antonin Morange (1895). De la psittacose ou infection spéciale '\n",
      "          'déterminée par les perruches. Médecine de Paris.\\n'\n",
      "          'Weekly Epidemiological Record. World Health Organization; IRIS',\n",
      "  'url': 'https://en.wikipedia.org/wiki/1929%E2%80%931930_psittacosis_pandemic'},\n",
      " {'article_title': 'COVID-19 pandemic',\n",
      "  'score': 7.26242,\n",
      "  'section_title': 'Data and graphs',\n",
      "  'text': 'COVID-19 pandemic,External links,Data and graphs\\n'\n",
      "          'Coronavirus disease (COVID-2019) situation reports and map by the '\n",
      "          'World Health Organization\\n'\n",
      "          'Coronavirus Resource Center, map, and historical data by Johns '\n",
      "          'Hopkins University\\n'\n",
      "          'Historical data about COVID-19 cases published by the European '\n",
      "          'Centre for Disease Prevention and Control (ECDC)\\n'\n",
      "          'World Travel Restrictions based on WFP data\\n'\n",
      "          'Coronavirus Observer based on Johns Hopkins University data\\n'\n",
      "          'COVID-19 coronavirus pandemic published by Worldometer\\n'\n",
      "          'Coronavirus Disease (COVID-19) Statistics and Research published by '\n",
      "          'Our World in Data\\n'\n",
      "          'COVID-19 Projections for many countries published by Institute for '\n",
      "          'Health Metrics and Evaluation\\n'\n",
      "          'Spreadsheets: Country comparisons\\n'\n",
      "          'Spreadsheets: Historical data with additional computations\\n'\n",
      "          'Epidemic Calculator\\n'\n",
      "          'Stat News COVID-19 Tracker',\n",
      "  'url': 'https://en.wikipedia.org/wiki/COVID-19_pandemic'}]\n"
     ]
    }
   ],
   "source": [
    "# print the three top results\n",
    "bool_responses = []\n",
    "pprint(top_three_results(bool_docs, responses=bool_responses))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Field boosting**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# boost title fields by a factor of 3\n",
    "boost_question = \"covid-19 pandemic\"\n",
    "boost_query_body = {\n",
    "    \"query\": {\n",
    "        \"multi_match\": {\n",
    "            \"query\": boost_question,\n",
    "            \"fields\": [ \"article_title^3\", \"section_title^3\" ] \n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/1784758995.py:2: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  boost_docs = client.search(body = boost_query_body, index=\"pandemics\", size=5)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'took': 25,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 131, 'relation': 'eq'},\n",
       "  'max_score': 27.540213,\n",
       "  'hits': [{'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': '-r9dr34BCwOBK0LKaHxc',\n",
       "    '_score': 27.540213,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'COVID-19',\n",
       "     'text': 'Pandemic,Current pandemics,COVID-19\\nA new strain of coronavirus was first identified in the city of Wuhan, Hubei province, China, in late December 2019. It has caused a cluster of cases of an acute respiratory disease, which is referred to as coronavirus disease 2019 (COVID-19). According to media reports, more than 200 countries and territories have been affected by COVID-19, with major outbreaks occurring in Brazil, Russia, India, Mexico, Peru, South Africa, Western Europe and the United States. On 11 March 2020, the World Health Organization characterized the spread of COVID-19 as a pandemic. As of 16 November 2020, the number of people infected with COVID-19 has reached 54,978,057 worldwide, of whom 38,243,617 have recovered. The death toll is 1,326,860. It is believed that these figures are understated as testing did not commence in the initial stages of the outbreak and many people infected by the virus have no or only mild symptoms and may not have been tested. Similarly, the number of recoveries may also be understated as tests are required before cases are officially recognised as recovered, and fatalities are sometimes attributed to other conditions. This was especially the case in large urban areas where a non-trivial number of patients died while in their private residences. It was later discovered that asymptomatic hypoxia due to COVID-19 pulmonary disease may be responsible for many such cases.',\n",
       "     'main_section': 'Current pandemics',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Current pandemics,COVID-19',\n",
       "     'section_number': 8}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'E79dr34BCwOBK0LKaH5l',\n",
       "    '_score': 27.540213,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'COVID-19',\n",
       "     'text': 'Science diplomacy and pandemics,COVID-19\\nOn December 31, 2019, authorities in Wuhan, China reported the first cases of an unknown pneumonia to the World Health Organization (WHO). Shortly after that, on January 4, 2020, the WHO reported on social media that there had been a group of pneumonia cases in Wuhan, China but that no deaths had been reported. The unknown pneumonia was a novel coronavirus, which they later named COVID-19. WHO quickly realized that it was a serious problem and issues between science and diplomacy on the topic of COVID-19 became evident. The Chinese government did not alert the WHO about the full extent of the seriousness of the outbreak.By the time the WHO realized that the problem was serious, the false information of a low mortality rate and its low risk to non-senior citizens had already spread very rapidly. The United States and the WHO tried to send scientists to help China fight COVID-19; however, China refused the help. China did eventually share the genome sequence of COVID-19 with the global community and later accepted another offer from the WHO to send a team of scientists to help. The lack of communication and the slow transmission of information between science and government caused failures to slow the spread of the virus in its early stages. To further complicate this stage, when scientists did give information or guidance, governments did not take it into proper consideration. As a result, by not acting fast enough, many people were infected and local economies suffered due to the slow decision to act in unison. The overall impact on society would have been less detrimental if governments had acted in collaboration with each other and scientists.Some progress has been made among the scientific community itself in creating transparency, so local scientists in each region have access to the information they need to fight COVID-19. For example, the global scientific community began making efforts to communicate and share relevant information to help combat COVID-19 without delay. The WHO reported that a global research roadmap has been created in March 2020 to fight COVID-19. Scientists are publishing research findings before they are peer-reviewed and sharing other discoveries without delay. Scientists and clinicians are working to fight the epidemic by sharing data through information sharing technologies using bioRxiv, medRxiv, ChemRxiv, and arXiv, which allows them to share information in real time. These tools are helping biopharma research and development, academic labs, government regulators, and the clinical community to speed up the process of testing new pathways to treating COVID-19 patients more efficiently. As one researcher explains, “We are experiencing a shift towards Open Science at a speed that was previously unthinkable. It began with the publication of the genetic sequence of COVID-19 by Chinese scientists in early January 2020 via GenBank – an open-access DNA database operated by the US National Center for Biotechnology Information”Additionally, The World Economic Forum has partnered with the WHO to mobilize and help supply the needed resources required to fight COVID-19 itself and the economic impact it is having. With the sharing of this information around the world by scientific communities, each individual country is better equipped with the knowledge they need to combat the pandemic and investigate possible treatments, testing, and vaccines. In the most current WHO published table of COVID-19 vaccines, there are 124 candidates that are in different stages of vaccine development.',\n",
       "     'main_section': 'COVID-19',\n",
       "     'article_title': 'Science diplomacy and pandemics',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Science_diplomacy_and_pandemics',\n",
       "     'page_id': 63936969,\n",
       "     'tags': 'Science diplomacy and pandemics,COVID-19',\n",
       "     'section_number': 4}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'n79dr34BCwOBK0LKaH1h',\n",
       "    '_score': 19.724483,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'COVID-19 (2019–2020)',\n",
       "     'text': 'Disease X,Candidates,COVID-19 (2019–2020)\\nFrom the outset of the COVID-19 pandemic, experts have speculated whether COVID-19 met the criteria to be Disease X. In early February 2020, Chinese virologist Shi Zhengli from the Wuhan Institute of Virology, wrote that the first Disease X is from a coronavirus. Later that month, Marion Koopmans, Head of Viroscience at Erasmus University Medical Center in Rotterdam, and a member of the WHO\\'s R&D Blueprint Special Advisory Group, wrote in scientific journal Cell, \"Whether it will be contained or not, this outbreak is rapidly becoming the first true pandemic challenge that fits the disease X category\". At the same time, Peter Daszak, also a member of the WHO\\'s R&D Blueprint, wrote in an opinion piece in the New York Times saying: \"In a nutshell, Covid-19 is Disease X\".',\n",
       "     'main_section': 'Candidates',\n",
       "     'article_title': 'Disease X',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Disease_X',\n",
       "     'page_id': 56809815,\n",
       "     'tags': 'Disease X,Candidates,COVID-19 (2019–2020)',\n",
       "     'section_number': 6}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'HL9dr34BCwOBK0LKaH5m',\n",
       "    '_score': 19.724483,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'COVID-19 outbreak 2020',\n",
       "     'text': 'Superspreading event,Superspreaders during outbreaks,COVID-19 outbreak 2020\\nThe South Korean spread of confirmed cases of SARS-CoV-2 infection jumped suddenly starting on 19–20 February 2020. On 19 February, the number of confirmed cases increased by 20. On 20 February, 58 or 70 new cases were confirmed, giving a total of 104 confirmed cases, according to the Centers for Disease Control and Prevention Korea (KCDC). According to Reuters, KCDC attributed the sudden jump to 70 cases linked to \"Patient 31\", who had participated in a gathering in Daegu at the Shincheonji Church of Jesus the Temple of the Tabernacle of the Testimony. On 20 February, the streets of Daegu were empty in reaction to the Shincheonji outbreak. A resident described the reaction, stating \"It\\'s like someone dropped a bomb in the middle of the city. It looks like a zombie apocalypse.\" On 21 February, the first death was reported. According to the mayor of Daegu, the number of suspected cases as of 21 February is 544 among 4,400 examined followers of the church. Later in the outbreak, in May, A 29-year-old man visited several Seoul nightclubs in one night and resulted in accumulated infections of at least 79 other people.A business conference in Boston (MA) from February 26–28 was a superspreading event. \\nIn New York, a lawyer contracted the illness then spread it to at least twenty other individuals in his community in New Rochelle, creating a cluster of cases that quickly passed 100, accounting for more than half of SARS2 coronavirus cases in the state during early March 2020. For comparison, the basic reproduction number of the virus, which is the average number of additional people that a single case will infect without any preventative measures, is between 1.4 and 3.9.In India, Baldev Singh, a preacher returned from Italy and Germany on 06 March resulting in 19 of his relatives having tested positive with 550 people having had direct contact with him. Because of this, India\\'s government fears an outbreak and quarantined on 27 March 2020 in the State of Punjab, 40,000 residents from 20 villages. While initially reported that Baldev Singh had ignored self-quarantine orders, this was disputed by his co-travellers who claimed no orders to quarantine were given by local authorities.A Tablighi Jamaat religious congregation that took place in Delhi\\'s Nizamuddin Markaz Mosque in early March 2020 was a coronavirus super-spreader event, with more than 4,000 confirmed cases and at least 27 deaths linked to the event reported across the country. Over 9,000 missionaries may have attended the congregation, with the majority being from various states of India, and 960 attendees from 40 foreign countries. On 18 April, 4,291 confirmed cases of COVID-19 linked to this event by the Union Health Ministry represented a third of all the confirmed cases of India. Around 40,000 people, including Tablighi Jamaat attendees and their contacts, were quarantined across the country.\\nOn 11 May 2020, it came to light that a worker at a fish processing plant in Tema, Ghana is believed to have infected over 500 other people with COVID-19.As of 18 July 2020, more than one thousand suspected superspreading events had been logged, for example a cluster of 187 people who were infected after eating at a Harper\\'s Restaurant and Brew Pub in East Lansing, Michigan.Several factors are identified as contributing to superspreading events with COVID-19: closed spaces with poor ventilation, crowds, and close contact settings (\"three Cs\").Statistical analyses of the frequency of coronavirus superspreading events, including SARS-CoV-2 and SARS, have shown that they correspond to fat-tailed events, indicating that they are extreme, but likely, occurrences.A SARS-CoV-2 superspreading events database maintained by a group of researchers at the London School of Hygiene and Tropical Medicine includes more than 1,600 superspreading events from around the world.',\n",
       "     'main_section': 'Superspreaders during outbreaks',\n",
       "     'article_title': 'Superspreading event',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Superspreading_event',\n",
       "     'page_id': 39958236,\n",
       "     'tags': 'Superspreading event,Superspreaders during outbreaks,COVID-19 outbreak 2020',\n",
       "     'section_number': 8}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Wb9dr34BCwOBK0LKaH1f',\n",
       "    '_score': 13.5429125,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Summary',\n",
       "     'text': \"COVID-19 pandemic,Summary\\nThe COVID-19 pandemic, also known as the coronavirus pandemic, is an ongoing pandemic of coronavirus disease 2019 (COVID-19) caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2), first identified in December 2019 in Wuhan, China. The World Health Organisation declared the outbreak a Public Health Emergency of International Concern in January 2020 and a pandemic in March 2020. As of 24 November 2020, more than 59.1 million cases have been confirmed, with more than 1.39 million deaths attributed to COVID-19.\\n\\nCommon symptoms include fever, cough, fatigue, breathing difficulties, and loss of smell and taste. Complications may include pneumonia and acute respiratory distress syndrome. The incubation period is typically around five days but may range from one to 14 days. There are several vaccine candidates in development, although none has completed clinical trials.  Although work is underway to develop drugs that inhibit the virus, the primary treatment is currently symptomatic.Recommended preventive measures include social distancing, wearing a face mask in public, ventilation and air-filtering, hand washing, covering one's mouth when sneezing or coughing, disinfecting surfaces, and monitoring and self-isolation for people exposed or symptomatic. Authorities worldwide have responded by implementing travel restrictions, lockdowns, workplace hazard controls, and facility closures. Many places have also worked to increase testing capacity and trace contacts of the infected.\\n\\nThe responses have caused global social and economic disruption, including the largest global recession since the Great Depression. It has led to the postponement or cancellation of events, widespread supply shortages exacerbated by panic buying, agricultural disruption and food shortages, and decreased emissions of pollutants and greenhouse gases. Educational institutions have been partially or fully closed. Misinformation has circulated through social media and mass media. There have been incidents of xenophobia and discrimination against Chinese people and against those perceived as being Chinese or as being from areas with high infection rates.\",\n",
       "     'main_section': 'Summary',\n",
       "     'article_title': 'COVID-19 pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/COVID-19_pandemic',\n",
       "     'page_id': 62750956,\n",
       "     'tags': 'COVID-19 pandemic,Summary',\n",
       "     'section_number': 0}}]}}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# submit the search query to elasticsearch\n",
    "boost_docs = client.search(body = boost_query_body, index=\"pandemics\", size=5)\n",
    "\n",
    "print(type(boost_docs))\n",
    "boost_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'article_title': 'Pandemic',\n",
      "  'score': 27.540213,\n",
      "  'section_title': 'COVID-19',\n",
      "  'text': 'Pandemic,Current pandemics,COVID-19\\n'\n",
      "          'A new strain of coronavirus was first identified in the city of '\n",
      "          'Wuhan, Hubei province, China, in late December 2019. It has caused '\n",
      "          'a cluster of cases of an acute respiratory disease, which is '\n",
      "          'referred to as coronavirus disease 2019 (COVID-19). According to '\n",
      "          'media reports, more than 200 countries and territories have been '\n",
      "          'affected by COVID-19, with major outbreaks occurring in Brazil, '\n",
      "          'Russia, India, Mexico, Peru, South Africa, Western Europe and the '\n",
      "          'United States. On 11 March 2020, the World Health Organization '\n",
      "          'characterized the spread of COVID-19 as a pandemic. As of 16 '\n",
      "          'November 2020, the number of people infected with COVID-19 has '\n",
      "          'reached 54,978,057 worldwide, of whom 38,243,617 have recovered. '\n",
      "          'The death toll is 1,326,860. It is believed that these figures are '\n",
      "          'understated as testing did not commence in the initial stages of '\n",
      "          'the outbreak and many people infected by the virus have no or only '\n",
      "          'mild symptoms and may not have been tested. Similarly, the number '\n",
      "          'of recoveries may also be understated as tests are required before '\n",
      "          'cases are officially recognised as recovered, and fatalities are '\n",
      "          'sometimes attributed to other conditions. This was especially the '\n",
      "          'case in large urban areas where a non-trivial number of patients '\n",
      "          'died while in their private residences. It was later discovered '\n",
      "          'that asymptomatic hypoxia due to COVID-19 pulmonary disease may be '\n",
      "          'responsible for many such cases.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Pandemic'},\n",
      " {'article_title': 'Science diplomacy and pandemics',\n",
      "  'score': 27.540213,\n",
      "  'section_title': 'COVID-19',\n",
      "  'text': 'Science diplomacy and pandemics,COVID-19\\n'\n",
      "          'On December 31, 2019, authorities in Wuhan, China reported the '\n",
      "          'first cases of an unknown pneumonia to the World Health '\n",
      "          'Organization (WHO). Shortly after that, on January 4, 2020, the WHO '\n",
      "          'reported on social media that there had been a group of pneumonia '\n",
      "          'cases in Wuhan, China but that no deaths had been reported. The '\n",
      "          'unknown pneumonia was a novel coronavirus, which they later named '\n",
      "          'COVID-19. WHO quickly realized that it was a serious problem and '\n",
      "          'issues between science and diplomacy on the topic of COVID-19 '\n",
      "          'became evident. The Chinese government did not alert the WHO about '\n",
      "          'the full extent of the seriousness of the outbreak.By the time the '\n",
      "          'WHO realized that the problem was serious, the false information of '\n",
      "          'a low mortality rate and its low risk to non-senior citizens had '\n",
      "          'already spread very rapidly. The United States and the WHO tried to '\n",
      "          'send scientists to help China fight COVID-19; however, China '\n",
      "          'refused the help. China did eventually share the genome sequence of '\n",
      "          'COVID-19 with the global community and later accepted another offer '\n",
      "          'from the WHO to send a team of scientists to help. The lack of '\n",
      "          'communication and the slow transmission of information between '\n",
      "          'science and government caused failures to slow the spread of the '\n",
      "          'virus in its early stages. To further complicate this stage, when '\n",
      "          'scientists did give information or guidance, governments did not '\n",
      "          'take it into proper consideration. As a result, by not acting fast '\n",
      "          'enough, many people were infected and local economies suffered due '\n",
      "          'to the slow decision to act in unison. The overall impact on '\n",
      "          'society would have been less detrimental if governments had acted '\n",
      "          'in collaboration with each other and scientists.Some progress has '\n",
      "          'been made among the scientific community itself in creating '\n",
      "          'transparency, so local scientists in each region have access to the '\n",
      "          'information they need to fight COVID-19. For example, the global '\n",
      "          'scientific community began making efforts to communicate and share '\n",
      "          'relevant information to help combat COVID-19 without delay. The WHO '\n",
      "          'reported that a global research roadmap has been created in March '\n",
      "          '2020 to fight COVID-19. Scientists are publishing research findings '\n",
      "          'before they are peer-reviewed and sharing other discoveries without '\n",
      "          'delay. Scientists and clinicians are working to fight the epidemic '\n",
      "          'by sharing data through information sharing technologies using '\n",
      "          'bioRxiv, medRxiv, ChemRxiv, and arXiv, which allows them to share '\n",
      "          'information in real time. These tools are helping biopharma '\n",
      "          'research and development, academic labs, government regulators, and '\n",
      "          'the clinical community to speed up the process of testing new '\n",
      "          'pathways to treating COVID-19 patients more efficiently. As one '\n",
      "          'researcher explains, “We are experiencing a shift towards Open '\n",
      "          'Science at a speed that was previously unthinkable. It began with '\n",
      "          'the publication of the genetic sequence of COVID-19 by Chinese '\n",
      "          'scientists in early January 2020 via GenBank – an open-access DNA '\n",
      "          'database operated by the US National Center for Biotechnology '\n",
      "          'Information”Additionally, The World Economic Forum has partnered '\n",
      "          'with the WHO to mobilize and help supply the needed resources '\n",
      "          'required to fight COVID-19 itself and the economic impact it is '\n",
      "          'having. With the sharing of this information around the world by '\n",
      "          'scientific communities, each individual country is better equipped '\n",
      "          'with the knowledge they need to combat the pandemic and investigate '\n",
      "          'possible treatments, testing, and vaccines. In the most current WHO '\n",
      "          'published table of COVID-19 vaccines, there are 124 candidates that '\n",
      "          'are in different stages of vaccine development.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Science_diplomacy_and_pandemics'},\n",
      " {'article_title': 'Disease X',\n",
      "  'score': 19.724483,\n",
      "  'section_title': 'COVID-19 (2019–2020)',\n",
      "  'text': 'Disease X,Candidates,COVID-19 (2019–2020)\\n'\n",
      "          'From the outset of the COVID-19 pandemic, experts have speculated '\n",
      "          'whether COVID-19 met the criteria to be Disease X. In early '\n",
      "          'February 2020, Chinese virologist Shi Zhengli from the Wuhan '\n",
      "          'Institute of Virology, wrote that the first Disease X is from a '\n",
      "          'coronavirus. Later that month, Marion Koopmans, Head of Viroscience '\n",
      "          'at Erasmus University Medical Center in Rotterdam, and a member of '\n",
      "          \"the WHO's R&D Blueprint Special Advisory Group, wrote in scientific \"\n",
      "          'journal Cell, \"Whether it will be contained or not, this outbreak '\n",
      "          'is rapidly becoming the first true pandemic challenge that fits the '\n",
      "          'disease X category\". At the same time, Peter Daszak, also a member '\n",
      "          \"of the WHO's R&D Blueprint, wrote in an opinion piece in the New \"\n",
      "          'York Times saying: \"In a nutshell, Covid-19 is Disease X\".',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Disease_X'}]\n"
     ]
    }
   ],
   "source": [
    "# print the three top results\n",
    "boost_responses = []\n",
    "pprint(top_three_results(boost_docs, responses=boost_responses))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Highlights**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a query body and a search string (...)\n",
    "hl_question = \"genetic mutations\"\n",
    "hl_query_body = {\n",
    "    \"query\": {\n",
    "        \"match\": { \"text\": hl_question }\n",
    "    },\n",
    "    \"highlight\": {\n",
    "        \"fields\": {\"text\": {\"number_of_fragments\" : 1, 'fragment_size':256}\n",
    "                  }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/4021062749.py:2: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  hl_docs = client.search(body = hl_query_body, index=\"pandemics\", size=5)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'dict'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'took': 44,\n",
       " 'timed_out': False,\n",
       " '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0},\n",
       " 'hits': {'total': {'value': 28, 'relation': 'eq'},\n",
       "  'max_score': 8.579811,\n",
       "  'hits': [{'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Z79dr34BCwOBK0LKaH5p',\n",
       "    '_score': 8.579811,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Genetic mutation',\n",
       "     'text': 'Virus,Microbiology,Genetic mutation\\nViruses undergo genetic change by several mechanisms. These include a process called antigenic drift where individual bases in the DNA or RNA mutate to other bases. Most of these point mutations are \"silent\"—they do not change the protein that the gene encodes—but others can confer evolutionary advantages such as resistance to antiviral drugs. Antigenic shift occurs when there is a major change in the genome of the virus. This can be a result of recombination or reassortment. When this happens with influenza viruses, pandemics might result. RNA viruses often exist as quasispecies or swarms of viruses of the same species but with slightly different genome nucleoside sequences. Such quasispecies are a prime target for natural selection.Segmented genomes confer evolutionary advantages; different strains of a virus with a segmented genome can shuffle and combine genes and produce progeny viruses (or offspring) that have unique characteristics. This is called reassortment or \\'viral sex\\'.Genetic recombination is the process by which a strand of DNA is broken and then joined to the end of a different DNA molecule. This can occur when viruses infect cells simultaneously and studies of viral evolution have shown that recombination has been rampant in the species studied. Recombination is common to both RNA and DNA viruses.',\n",
       "     'main_section': 'Microbiology',\n",
       "     'article_title': 'Virus',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Virus',\n",
       "     'page_id': 19167679,\n",
       "     'tags': 'Virus,Microbiology,Genetic mutation',\n",
       "     'section_number': 10},\n",
       "    'highlight': {'text': ['Virus,Microbiology,<em>Genetic</em> mutation\\nViruses undergo <em>genetic</em> change by several mechanisms. These include a process called antigenic drift where individual bases in the DNA or RNA mutate to other bases.']}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Br9dr34BCwOBK0LKaH1c',\n",
       "    '_score': 8.050789,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Viral hemorrhagic fevers',\n",
       "     'text': 'Pandemic,Concerns about future pandemics,Concerning diseases,Viral hemorrhagic fevers\\nViral hemorrhagic fevers such as Ebola virus disease, Lassa fever, Rift Valley fever, Marburg virus disease, Bolivian hemorrhagic fever and Crimean-Congo hemorrhagic fever are highly contagious and deadly diseases, with the theoretical potential to become pandemics. Their ability to spread efficiently enough to cause a pandemic is limited, however, as transmission of these viruses requires close contact with the infected vector, and the vector has only a short time before death or serious illness. Furthermore, the short time between a vector becoming infectious and the onset of symptoms allows medical professionals to quickly quarantine vectors, and prevent them from carrying the pathogen elsewhere. Genetic mutations could occur, which could elevate their potential for causing widespread harm; thus close observation by contagious disease specialists is merited.',\n",
       "     'main_section': 'Concerns about future pandemics',\n",
       "     'article_title': 'Pandemic',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic',\n",
       "     'page_id': 24255,\n",
       "     'tags': 'Pandemic,Concerns about future pandemics,Concerning diseases,Viral hemorrhagic fevers',\n",
       "     'section_number': 26},\n",
       "    'highlight': {'text': ['<em>Genetic</em> <em>mutations</em> could occur, which could elevate their potential for causing widespread harm; thus close observation by contagious disease specialists is merited.']}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'Yr9dr34BCwOBK0LKaH5p',\n",
       "    '_score': 7.2686186,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Life properties',\n",
       "     'text': 'Virus,Microbiology,Life properties\\nScientific opinions differ on whether viruses are a form of life or organic structures that interact with living organisms. They have been described as \"organisms at the edge of life\", since they resemble organisms in that they possess genes, evolve by natural selection, and reproduce by creating multiple copies of themselves through self-assembly. Although they have genes, they do not have a cellular structure, which is often seen as the basic unit of life. Viruses do not have their own metabolism and require a host cell to make new products. They therefore cannot naturally reproduce outside a host cell—although bacterial species such as rickettsia and chlamydia are considered living organisms despite the same limitation. Accepted forms of life use cell division to reproduce, whereas viruses spontaneously assemble within cells. They differ from autonomous growth of crystals as they inherit genetic mutations while being subject to natural selection. Virus self-assembly within host cells has implications for the study of the origin of life, as it lends further credence to the hypothesis that life could have started as self-assembling organic molecules.',\n",
       "     'main_section': 'Microbiology',\n",
       "     'article_title': 'Virus',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Virus',\n",
       "     'page_id': 19167679,\n",
       "     'tags': 'Virus,Microbiology,Life properties',\n",
       "     'section_number': 5},\n",
       "    'highlight': {'text': ['They differ from autonomous growth of crystals as they inherit <em>genetic</em> <em>mutations</em> while being subject to natural selection.']}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': 'fr9dr34BCwOBK0LKaH5q',\n",
       "    '_score': 4.658639,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Role in evolution',\n",
       "     'text': 'Virus,Role in evolution,Microbiology\\nViruses are an important natural means of transferring genes between different species, which increases genetic diversity and drives evolution. It is thought that viruses played a central role in early evolution, before the diversification of the last universal common ancestor into bacteria, archaea and eukaryotes. Viruses are still one of the largest reservoirs of unexplored genetic diversity on Earth.',\n",
       "     'main_section': 'Role in evolution',\n",
       "     'article_title': 'Virus',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Virus',\n",
       "     'page_id': 19167679,\n",
       "     'tags': 'Virus,Role in evolution,Microbiology',\n",
       "     'section_number': 33},\n",
       "    'highlight': {'text': ['Viruses are still one of the largest reservoirs of unexplored <em>genetic</em> diversity on Earth.']}},\n",
       "   {'_index': 'pandemics',\n",
       "    '_type': '_doc',\n",
       "    '_id': '379dr34BCwOBK0LKaH1k',\n",
       "    '_score': 4.5918846,\n",
       "    '_ignored': ['text.keyword'],\n",
       "    '_source': {'section_title': 'Pathogen detection and prediction',\n",
       "     'text': 'Pandemic prevention,Measures,Technologies,Pathogen detection and prediction\\nIn a 2012 study it is claimed that \"new mathematical modelling, diagnostic, communications, and informatics technologies can identify and report hitherto unknown  microbes in other species, and thus new risk assessment approaches are needed to identify microbes most likely to cause human disease\". The study investigates challenges in moving the global pandemic strategy from response to pre-emption. Some scientists are screening blood samples from wildlife for new viruses. The international Global Virome Project (GVP) aims to identify the causes of fatal new diseases before emergence in human hosts by genetically characterizing viruses found in wild animals. The project aims to enlist an international network of scientists to collect hundreds of thousands of viruses, map their genomes, characterize and risk-stratify them to identify which ones to pay attention to. However, some infectious disease experts have criticized the project as too broad and expensive due to limited global scientific and financial resources and because only a small percentage of the world\\'s zoonotic viruses may cross into humans and pose a threat. They argue for prioritizing rapidly detecting diseases when they cross into humans and an improving the understanding of their mechanisms. A successful prevention of a pandemic from specific viruses may also require ensuring that it does not re-emerge – for instance by sustaining itself in domestic animals.Pathogen detection mechanisms may allow the construction of an early warning system which could make use of artificial intelligence surveillance and outbreak investigation. Edward Rubin notes that after sufficient data has been gathered artificial intelligence could be used to identify common features and develop countermeasures and vaccines against whole categories of viruses. It might be possible to predict viral evolution using machine learning. In April 2020 it was reported that researchers developed a predictive algorithm which can show in visualizations how combinations of genetic mutations can make proteins highly effective or ineffective in organisms – including for viral evolution for viruses like SARS-CoV-2. An artificial \"global immune system\"-like technological system that includes pathogen detection may be able to substantially reduce the time required to take on a biothreat agent. A system of that sort would also include a network of well-trained epidemiologists who could be rapidly deployed to investigate and contain an outbreak.Funding for the United States\\' PREDICT government research program that sought to identify animal pathogens that might infect humans and to prevent new pandemics was cut in 2019. Funding for United States\\' CDC programs that trained workers in outbreak detection and strengthened laboratory and emergency response systems in countries where disease risks are greatest to stop outbreaks at the source was cut by 80% in 2018.Despite recent advances in pandemic modeling, experts using mostly experience and intuition are still more accurate in predicting the spread of disease than strictly mathematical models.',\n",
       "     'main_section': 'Measures',\n",
       "     'article_title': 'Pandemic prevention',\n",
       "     'source_url': 'https://en.wikipedia.org/wiki/Pandemic_prevention',\n",
       "     'page_id': 63478457,\n",
       "     'tags': 'Pandemic prevention,Measures,Technologies,Pathogen detection and prediction',\n",
       "     'section_number': 5},\n",
       "    'highlight': {'text': ['In April 2020 it was reported that researchers developed a predictive algorithm which can show in visualizations how combinations of <em>genetic</em> <em>mutations</em> can make proteins highly effective or ineffective in organisms – including for viral evolution for viruses']}}]}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# submit the search query to elasticsearch\n",
    "hl_docs = client.search(body = hl_query_body, index=\"pandemics\", size=5)\n",
    "\n",
    "print(type(hl_docs))\n",
    "hl_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'article_title': 'Virus',\n",
      "  'score': 8.579811,\n",
      "  'section_title': 'Genetic mutation',\n",
      "  'text': 'Virus,Microbiology,Genetic mutation\\n'\n",
      "          'Viruses undergo genetic change by several mechanisms. These include '\n",
      "          'a process called antigenic drift where individual bases in the DNA '\n",
      "          'or RNA mutate to other bases. Most of these point mutations are '\n",
      "          '\"silent\"—they do not change the protein that the gene encodes—but '\n",
      "          'others can confer evolutionary advantages such as resistance to '\n",
      "          'antiviral drugs. Antigenic shift occurs when there is a major '\n",
      "          'change in the genome of the virus. This can be a result of '\n",
      "          'recombination or reassortment. When this happens with influenza '\n",
      "          'viruses, pandemics might result. RNA viruses often exist as '\n",
      "          'quasispecies or swarms of viruses of the same species but with '\n",
      "          'slightly different genome nucleoside sequences. Such quasispecies '\n",
      "          'are a prime target for natural selection.Segmented genomes confer '\n",
      "          'evolutionary advantages; different strains of a virus with a '\n",
      "          'segmented genome can shuffle and combine genes and produce progeny '\n",
      "          'viruses (or offspring) that have unique characteristics. This is '\n",
      "          \"called reassortment or 'viral sex'.Genetic recombination is the \"\n",
      "          'process by which a strand of DNA is broken and then joined to the '\n",
      "          'end of a different DNA molecule. This can occur when viruses infect '\n",
      "          'cells simultaneously and studies of viral evolution have shown that '\n",
      "          'recombination has been rampant in the species studied. '\n",
      "          'Recombination is common to both RNA and DNA viruses.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Virus'},\n",
      " {'article_title': 'Pandemic',\n",
      "  'score': 8.050789,\n",
      "  'section_title': 'Viral hemorrhagic fevers',\n",
      "  'text': 'Pandemic,Concerns about future pandemics,Concerning diseases,Viral '\n",
      "          'hemorrhagic fevers\\n'\n",
      "          'Viral hemorrhagic fevers such as Ebola virus disease, Lassa fever, '\n",
      "          'Rift Valley fever, Marburg virus disease, Bolivian hemorrhagic '\n",
      "          'fever and Crimean-Congo hemorrhagic fever are highly contagious and '\n",
      "          'deadly diseases, with the theoretical potential to become '\n",
      "          'pandemics. Their ability to spread efficiently enough to cause a '\n",
      "          'pandemic is limited, however, as transmission of these viruses '\n",
      "          'requires close contact with the infected vector, and the vector has '\n",
      "          'only a short time before death or serious illness. Furthermore, the '\n",
      "          'short time between a vector becoming infectious and the onset of '\n",
      "          'symptoms allows medical professionals to quickly quarantine '\n",
      "          'vectors, and prevent them from carrying the pathogen elsewhere. '\n",
      "          'Genetic mutations could occur, which could elevate their potential '\n",
      "          'for causing widespread harm; thus close observation by contagious '\n",
      "          'disease specialists is merited.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Pandemic'},\n",
      " {'article_title': 'Virus',\n",
      "  'score': 7.2686186,\n",
      "  'section_title': 'Life properties',\n",
      "  'text': 'Virus,Microbiology,Life properties\\n'\n",
      "          'Scientific opinions differ on whether viruses are a form of life or '\n",
      "          'organic structures that interact with living organisms. They have '\n",
      "          'been described as \"organisms at the edge of life\", since they '\n",
      "          'resemble organisms in that they possess genes, evolve by natural '\n",
      "          'selection, and reproduce by creating multiple copies of themselves '\n",
      "          'through self-assembly. Although they have genes, they do not have a '\n",
      "          'cellular structure, which is often seen as the basic unit of life. '\n",
      "          'Viruses do not have their own metabolism and require a host cell to '\n",
      "          'make new products. They therefore cannot naturally reproduce '\n",
      "          'outside a host cell—although bacterial species such as rickettsia '\n",
      "          'and chlamydia are considered living organisms despite the same '\n",
      "          'limitation. Accepted forms of life use cell division to reproduce, '\n",
      "          'whereas viruses spontaneously assemble within cells. They differ '\n",
      "          'from autonomous growth of crystals as they inherit genetic '\n",
      "          'mutations while being subject to natural selection. Virus '\n",
      "          'self-assembly within host cells has implications for the study of '\n",
      "          'the origin of life, as it lends further credence to the hypothesis '\n",
      "          'that life could have started as self-assembling organic molecules.',\n",
      "  'url': 'https://en.wikipedia.org/wiki/Virus'}]\n"
     ]
    }
   ],
   "source": [
    "# print the three top results\n",
    "hl_responses = []\n",
    "pprint(top_three_results(hl_docs, responses=hl_responses))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Reranking with BERT**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Boolean query**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using a boolean query to exclude irrelevant sections\n",
    "def boolean_query(exclude_sections, query, index=\"\"):\n",
    "\n",
    "    bool_query_body = {\n",
    "      \"query\": {\n",
    "        \"bool\": {\n",
    "            \"should\": \n",
    "            { \"match\": {\"text\": query}},\n",
    "            \"must_not\": {\n",
    "                \"terms\": {\"section_title.keyword\": exclude_sections}\n",
    "          },\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "\n",
    "    # submit a search query to ElasticSearch\n",
    "    bool_docs = client.search(body = bool_query_body, index=index, size=50)\n",
    "    \n",
    "    # reshape search results to prepare them for sentence embeddings\n",
    "    texts = []\n",
    "    section_titles = []\n",
    "    article_titles = []\n",
    "    score = []\n",
    "    \n",
    "    for doc in bool_docs[\"hits\"][\"hits\"]:\n",
    "        texts.append(doc[\"_source\"][\"text\"])\n",
    "        section_titles.append(doc[\"_source\"][\"section_title\"])\n",
    "        article_titles.append(doc[\"_source\"][\"article_title\"])\n",
    "        score.append(doc[\"_score\"])\n",
    "    \n",
    "    return texts, article_titles, section_titles, score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2272/2873066368.py:17: DeprecationWarning: The 'body' parameter is deprecated for the 'search' API and will be removed in a future version. Instead use API parameters directly. See https://github.com/elastic/elasticsearch-py/issues/1698 for more information\n",
      "  bool_docs = client.search(body = bool_query_body, index=index, size=50)\n"
     ]
    }
   ],
   "source": [
    "query = \"what is desease X?\"\n",
    "exclude_sections = [\"See also\", \"Further reading\", \"Data and graphs\", \"Medical journals\", \"External links\"]\n",
    "\n",
    "results = boolean_query(exclude_sections=exclude_sections,\n",
    "                       index=\"pandemics\",\n",
    "                       query=query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Disease X', 'Disease X', 'Disease X', 'Disease X', 'Disease X', 'Disease X', 'Johns Hopkins Center for Health Security', 'Disease X', 'Disease X', 'Disease X', 'Swine influenza', 'Pandemic severity index', 'Virus', 'Targeted immunization strategies', 'Cholera', 'Plague of Cyprian', 'Pandemic severity index', 'Cholera', 'Cholera', 'Johns Hopkins Center for Health Security', 'Crimson Contagion', 'HIV/AIDS', 'HIV/AIDS', 'Superspreading event', 'Pandemic', 'Basic reproduction number', 'COVID-19 pandemic', 'COVID-19 pandemic', 'Superspreading event', 'Epidemiology of HIV/AIDS', 'Swine influenza', 'COVID-19 pandemic', 'COVID-19 pandemic', 'Pandemic prevention', 'Epidemiology of HIV/AIDS', 'Pandemic', 'Pandemic', 'HIV/AIDS', 'Virus', 'Virus', 'Swine influenza', 'Cholera', 'HIV/AIDS', 'Viral load', 'Cholera', 'COVID-19 pandemic', 'Virus', 'Virus', 'HIV/AIDS', 'HIV/AIDS']\n"
     ]
    }
   ],
   "source": [
    "# print the titles (in descending order of elasticsearch's tfidf_scores )\n",
    "print(results[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sentence embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "elasticsearch uses tfidf_scores to rank the search results. This causes\n",
    "documents to be prioritized by the frequency of keywords in the documents \n",
    "normalized by the total number of terms. Hence, the results tend to be shorter \n",
    "documents rich in terms often receiving a higher relevance score compared to \n",
    "more descriptive, longer, and more information-rich documents. To improve the \n",
    "quality sentence embeddings for the results returned by elasticsearch\n",
    "sentence-transformers BERT models finetuned for semantic text search will be\n",
    "computed\n",
    "'''\n",
    "\n",
    "def compute_embeddings(query, results, model, top_k=10):\n",
    "    texts = results[0]\n",
    "    article_titles = results[1]\n",
    "    section_titles = results[2]\n",
    "    \n",
    "    emb = SentenceTransformer(model)\n",
    "    corpus_emb = emb.encode(texts, convert_to_tensor=True)\n",
    "    query_emb = emb.encode(query, convert_to_tensor=True)\n",
    "    # returns a list of dictionaries with the keys 'corpus_id' and 'score',\n",
    "    # sorted by decreasing cosine similarity scores\n",
    "    reranked_results = util.semantic_search(query_emb, corpus_emb, top_k=top_k)[0]\n",
    "    \n",
    "    rr_results_list = []\n",
    "    for item in reranked_results:\n",
    "        idx = item['corpus_id']\n",
    "        rr_results_dict = {\n",
    "            'bert_score': item['score'],\n",
    "            'article_title': article_titles[idx],\n",
    "            'section_title': section_titles[idx],\n",
    "            'text': texts[idx]\n",
    "        }\n",
    "        rr_results_list.append(rr_results_dict)\n",
    "    \n",
    "    return rr_results_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'article_title': 'Disease X',\n",
      "  'bert_score': 0.29652318358421326,\n",
      "  'section_title': 'Summary',\n",
      "  'text': 'Disease X,Summary\\n'\n",
      "          'Disease X is a placeholder name that was adopted by the World '\n",
      "          'Health Organization (WHO) in February 2018 on their shortlist of '\n",
      "          'blueprint priority diseases to represent a hypothetical, unknown '\n",
      "          'pathogen that could cause a future epidemic. The WHO adopted the '\n",
      "          'placeholder name to ensure that their planning was sufficiently '\n",
      "          'flexible to adapt to an unknown pathogen (e.g. broader vaccines and '\n",
      "          'manufacturing facilities).  Director of the US National Institute '\n",
      "          'of Allergy and Infectious Diseases Anthony Fauci stated that the '\n",
      "          'concept of Disease X would encourage WHO projects to focus their '\n",
      "          'research efforts on entire classes of viruses (e.g. flaviviruses), '\n",
      "          'instead of just individual strains (e.g. zika virus), thus '\n",
      "          'improving WHO capability to respond to unforeseen strains.  In '\n",
      "          \"2020, it was speculated, including among some of the WHO's own \"\n",
      "          'expert advisors, that COVID-19, caused by SARS-CoV-2 virus strain, '\n",
      "          'met the requirements to be the first Disease X.'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.27254217863082886,\n",
      "  'section_title': 'Rationale',\n",
      "  'text': 'Disease X,Rationale\\n'\n",
      "          'In May 2015, in pandemic preparations prior to the COVID-19 '\n",
      "          'pandemic, the WHO was asked by member organizations to create an '\n",
      "          '\"R&D Blueprint for Action to Prevent Epidemics\" to generate ideas '\n",
      "          'that would reduce the time lag between the identification of viral '\n",
      "          'outbreaks and the approval of vaccines/treatments, to stop the '\n",
      "          'outbreaks from turning into a \"public health emergency\". The focus '\n",
      "          'was to be on the most serious emerging infectious diseases (EIDs) '\n",
      "          'for which few preventive options were available. A group of global '\n",
      "          'experts, the \"R&D Blueprint Scientific Advisory Group\", was '\n",
      "          'assembled by the WHO to draft a shortlist of less than ten '\n",
      "          '\"blueprint priority diseases\".Since 2015, the shortlist of less '\n",
      "          'than 10 EIDs has been updated annually and has consistently '\n",
      "          'included widely known names such as Ebola, Zika and SARS (e.g. '\n",
      "          'cause of large-scale infections), and more geographically specific '\n",
      "          'names such as Lassa fever, Marburg virus, Rift Valley fever, and '\n",
      "          'Nipah virus.In February 2018, after the \"2018 R&D Blueprint\" '\n",
      "          'meeting in Geneva, the WHO added Disease X to the shortlist as a '\n",
      "          'placeholder for a \"knowable unknown\" pathogen.  The Disease X '\n",
      "          'placeholder acknowledged the potential for a future epidemic that '\n",
      "          'could be caused by an unknown pathogen, and by its inclusion, '\n",
      "          'challenged the WHO to ensure their planning and capabilities were '\n",
      "          'flexible enough to adapt to such an event.At the 2018 announcement '\n",
      "          'of the updated shortlist of blueprint priority diseases, the WHO '\n",
      "          'said: \"Disease X represents the knowledge that a serious '\n",
      "          'international epidemic could be caused by a pathogen currently '\n",
      "          'unknown to cause human disease\". John-Arne Røttingen, of the R&D '\n",
      "          'Blueprint Special Advisory Group, said: \"History tells us that it '\n",
      "          'is likely the next big outbreak will be something we have not seen '\n",
      "          'before\", and \"It may seem strange to be adding an \\'X\\' but the '\n",
      "          'point is to make sure we prepare and plan flexibly in terms of '\n",
      "          \"vaccines and diagnostic tests. We want to see 'plug and play' \"\n",
      "          'platforms developed which will work for any, or a wide number of '\n",
      "          'diseases; systems that will allow us to create countermeasures at '\n",
      "          'speed\". US expert Anthony Fauci said: \"WHO recognizes it must '\n",
      "          '\\'nimbly move\\' and this involves creating platform technologies\", '\n",
      "          'and that to develop such platforms, WHO would have to research '\n",
      "          'entire classes of viruses, highlighting flaviviruses. He added: \"If '\n",
      "          'you develop an understanding of the commonalities of those, you can '\n",
      "          'respond more rapidly\".'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.2663199305534363,\n",
      "  'section_title': 'Adoption',\n",
      "  'text': 'Disease X,Adoption\\n'\n",
      "          \"Jonathan D. Quick, author of End of Epidemics, described the WHO's \"\n",
      "          'act of naming Disease X as \"wise in terms of communicating risk\", '\n",
      "          'saying \"panic and complacency are  the hallmarks of the world\\'s '\n",
      "          'response to infectious diseases, with complacency currently in the '\n",
      "          'ascendance\".  Women\\'s Health wrote that the establishment of the '\n",
      "          'term \"might seem like an uncool move designed to incite panic\" but '\n",
      "          'that the whole purpose of including it on the list was to \"get it '\n",
      "          'on people’s radars\".Richard Hatchett of the Coalition for Epidemic '\n",
      "          'Preparedness Innovations (CEPI), wrote \"It might sound like science '\n",
      "          'fiction, but Disease X is something we must prepare for\", noting '\n",
      "          'that despite the success in controlling the 2014 Western African '\n",
      "          'Ebola virus epidemic, strains of the disease had returned in 2018. '\n",
      "          'In February 2019, CEPI announced funding of US$34 million to the '\n",
      "          'German-based CureVac biopharmaceutical company to develop an \"RNA '\n",
      "          'Printer prototype\", that CEPI said could \"prepare for rapid '\n",
      "          'response to unknown pathogens (i.e., Disease X)\".Parallels were '\n",
      "          'drawn with the efforts by the United States Agency for '\n",
      "          'International Development (USAID) and their PREDICT programme, '\n",
      "          'which was designed to act as an early warning pandemic system, by '\n",
      "          'sourcing and researching animal viruses in particular \"hot spots\" '\n",
      "          'of animal-human interaction.In September 2019, The Daily Telegraph '\n",
      "          'reported on how Public Health England (PHE) had launched its own '\n",
      "          'investigation for a potential Disease X in the United Kingdom from '\n",
      "          'the diverse range of diseases reported in their health system; they '\n",
      "          'noted that 12 novel diseases and/or viruses had been recorded by '\n",
      "          \"PHE in the last decade.In October 2019 in New York, the WHO's \"\n",
      "          'Health Emergencies Programme ran a \"Disease X dummy run\" to '\n",
      "          'simulate a global pandemic by a Disease X, in order for its 150 '\n",
      "          'participants from various world health agencies and public health '\n",
      "          'systems to better prepare and share ideas and observations for '\n",
      "          'combatting such an eventuality.In March 2020, The Lancet Infectious '\n",
      "          'Diseases published a paper titled \"Disease X: accelerating the '\n",
      "          'development of medical countermeasures for the next pandemic\", '\n",
      "          'which expanded the term to include Pathogen X (the pathogen that '\n",
      "          'leads to Disease X), and identified areas of product development '\n",
      "          'and international coordination that would help in combatting any '\n",
      "          'future Disease X.In April 2020, The Daily Telegraph described '\n",
      "          'remdesivir, a drug being trialed to combat coronavirus, as an '\n",
      "          'anti-viral that Gilead Sciences started working on a decade '\n",
      "          'previously to treat a future Disease X.'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.24858306348323822,\n",
      "  'section_title': 'COVID-19 (2019–2020)',\n",
      "  'text': 'Disease X,Candidates,COVID-19 (2019–2020)\\n'\n",
      "          'From the outset of the COVID-19 pandemic, experts have speculated '\n",
      "          'whether COVID-19 met the criteria to be Disease X. In early '\n",
      "          'February 2020, Chinese virologist Shi Zhengli from the Wuhan '\n",
      "          'Institute of Virology, wrote that the first Disease X is from a '\n",
      "          'coronavirus. Later that month, Marion Koopmans, Head of Viroscience '\n",
      "          'at Erasmus University Medical Center in Rotterdam, and a member of '\n",
      "          \"the WHO's R&D Blueprint Special Advisory Group, wrote in scientific \"\n",
      "          'journal Cell, \"Whether it will be contained or not, this outbreak '\n",
      "          'is rapidly becoming the first true pandemic challenge that fits the '\n",
      "          'disease X category\". At the same time, Peter Daszak, also a member '\n",
      "          \"of the WHO's R&D Blueprint, wrote in an opinion piece in the New \"\n",
      "          'York Times saying: \"In a nutshell, Covid-19 is Disease X\".'},\n",
      " {'article_title': 'Superspreading event',\n",
      "  'bert_score': 0.23892180621623993,\n",
      "  'section_title': 'Defining a superspreader event',\n",
      "  'text': 'Superspreading event,Defining a superspreader event\\n'\n",
      "          'Although loose definitions of superspreader events exist, some '\n",
      "          'effort has been made at defining what qualifies as a superspreader '\n",
      "          'event (SSEV). Lloyd-Smith et al. (2005) define a protocol to '\n",
      "          'identify a superspreader event as follows:\\n'\n",
      "          'estimate the effective reproductive number, R, for the disease and '\n",
      "          'population in question;\\n'\n",
      "          'construct a Poisson distribution with mean R, representing the '\n",
      "          'expected range of Z due to stochasticity without individual '\n",
      "          'variation;\\n'\n",
      "          'define an SSEV as any infected person who infects more than Z(n) '\n",
      "          'others, where Z(n) is the nth percentile of the Poisson(R) '\n",
      "          'distribution.This protocol defines a 99th-percentile SSEV as a case '\n",
      "          'which causes more infections than would occur in 99% of infectious '\n",
      "          'histories in a homogeneous population.During the SARS-CoV-1 '\n",
      "          '2002–2004 SARS outbreak from China, epidemiologists defined a '\n",
      "          'superspreader as an individual with at least eight transmissions of '\n",
      "          'the disease.Superspreaders may or may not show any symptoms of the '\n",
      "          'disease.Writing in Quillette on 23 April 2020 and again a week '\n",
      "          'later in the National Post, Jonathan Kay noticed that in the case '\n",
      "          'of the SARS‑CoV‑2 virus outbreak from China which subsequently '\n",
      "          'turned into the COVID-19 pandemic:\\n'\n",
      "          'Putting aside hospitals, private residences and old-age homes, '\n",
      "          'almost all of these superspreader events (SSEVs) took place in the '\n",
      "          'context of (1) parties, (2) face-to-face professional networking '\n",
      "          'events and meetings, (3) religious gatherings, (4) sports events, '\n",
      "          '(5) meat-processing facilities, (6) ships at sea, (7) singing '\n",
      "          'groups, and, yes, (8) funerals.'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.21732479333877563,\n",
      "  'section_title': 'Synthetic viruses / bioweapons',\n",
      "  'text': 'Disease X,Candidates,Synthetic viruses / bioweapons\\n'\n",
      "          'At the 2018 announcement of the updated shortlist of blueprint '\n",
      "          'priority diseases, the media speculated that a future Disease X '\n",
      "          'could be created intentionally as a biological weapon.  In 2018, '\n",
      "          'WHO R&D Blueprint Special Advisor Group member Røttingen was '\n",
      "          'questioned about the potential of Disease X to come from the '\n",
      "          'ability of gene-editing technology to produce synthetic viruses '\n",
      "          '(e.g. the 2017 synthesis of Orthopoxvirus in Canada was cited), '\n",
      "          'which could be released through an accident or even an act of '\n",
      "          'terror. Røttingen said it was unlikely that a future Disease X '\n",
      "          'would originate from a synthetic virus or a bio-weapon. However, he '\n",
      "          'noted the seriousness of such an event,  saying, \"Synthetic biology '\n",
      "          'allows for the creation of deadly new viruses. It is also the case '\n",
      "          'that where you have a new disease there is no resistance in the '\n",
      "          'population and that means it can spread fast\".'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.21594154834747314,\n",
      "  'section_title': 'Bacterial infection',\n",
      "  'text': 'Disease X,Candidates,Bacterial infection\\n'\n",
      "          'In September 2019, Public Health England (PHE) reported that the '\n",
      "          'increasing antibiotic resistance of bacteria, even to \"last-resort\" '\n",
      "          'antibiotics such as carbapenems and colistin, could also turn into '\n",
      "          'a potential Disease X, citing the antibiotic resistance in '\n",
      "          'gonorrhea as an example.'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.21330812573432922,\n",
      "  'section_title': 'In popular culture',\n",
      "  'text': 'Disease X,In popular culture,Candidates\\n'\n",
      "          'In 2018, the Museum of London ran an exhibition titled \"Disease X: '\n",
      "          'London\\'s next epidemic?\", hosted for the centenary of the Spanish '\n",
      "          'flu epidemic from 1918.The term features in the title of several '\n",
      "          'fiction books that involve global pandemic diseases, such as '\n",
      "          'Disease  (2020), and Disease X: The Outbreak (2019).'},\n",
      " {'article_title': 'Basic reproduction number',\n",
      "  'bert_score': 0.19622772932052612,\n",
      "  'section_title': 'Summary',\n",
      "  'text': 'Basic reproduction number,Summary\\n'\n",
      "          'In epidemiology, the basic reproduction number, or basic '\n",
      "          'reproductive number (sometimes called basic reproduction ratio or '\n",
      "          'basic reproductive rate), denoted \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '   (pronounced R nought or R zero), of an infection can be thought '\n",
      "          'of as the expected number of cases directly generated by one case '\n",
      "          'in a population where all individuals are susceptible to infection. '\n",
      "          'The definition describes the state where no other individuals are '\n",
      "          'infected or immunized (naturally or through vaccination). Some '\n",
      "          'definitions, such as that of the Australian Department of Health, '\n",
      "          'add absence of \"any deliberate intervention in disease '\n",
      "          'transmission\". The basic reproduction number is not to be confused '\n",
      "          'with the effective reproduction number \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        R\\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R}\\n'\n",
      "          '   (usually written \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            t\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{t}}\\n'\n",
      "          '   [t for time], sometimes \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            e\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{e}}\\n'\n",
      "          '  ), which is the number of cases generated in the current state of '\n",
      "          'a population, which does not have to be the uninfected state. Also, '\n",
      "          'it is important to note that \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '   is a dimensionless number and not a rate, which would have units '\n",
      "          'of time−1, or units of time like doubling time.\\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '   is not a biological constant for a pathogen as it is also '\n",
      "          'affected by other factors such as environmental conditions and the '\n",
      "          'behaviour of the infected population. Furthermore \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '   values are usually estimated from mathematical models, and the '\n",
      "          'estimated values are dependent on the model used and values of '\n",
      "          'other parameters. Thus values given in the literature only make '\n",
      "          'sense in the given context and it is recommended not to use '\n",
      "          'obsolete values or compare values based on different models. \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '   does not by itself give an estimate of how fast an infection '\n",
      "          'spreads in the population.\\n'\n",
      "          'The most important uses of \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '   are determining if an emerging infectious disease can spread in '\n",
      "          'a population and determining what proportion of the population '\n",
      "          'should be immunized through vaccination to eradicate a disease. In '\n",
      "          'commonly used infection models, when \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '        >\\n'\n",
      "          '        1\\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}>1}\\n'\n",
      "          '   the infection will be able to start spreading in a population, '\n",
      "          'but not if \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '        <\\n'\n",
      "          '        1\\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}<1}\\n'\n",
      "          '  . Generally, the larger the value of \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle R_{0}}\\n'\n",
      "          '  , the harder it is to control the epidemic. For simple models, '\n",
      "          'the proportion of the population that needs to be effectively '\n",
      "          'immunized (meaning not susceptible to infection) to prevent '\n",
      "          'sustained spread of the infection has to be larger than \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        1\\n'\n",
      "          '        −\\n'\n",
      "          '        1\\n'\n",
      "          '        \\n'\n",
      "          '          /\\n'\n",
      "          '        \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle 1-1/R_{0}}\\n'\n",
      "          '  . Conversely, the proportion of the population that remains '\n",
      "          'susceptible to infection in the endemic equilibrium is \\n'\n",
      "          '  \\n'\n",
      "          '    \\n'\n",
      "          '      \\n'\n",
      "          '        1\\n'\n",
      "          '        \\n'\n",
      "          '          /\\n'\n",
      "          '        \\n'\n",
      "          '        \\n'\n",
      "          '          R\\n'\n",
      "          '          \\n'\n",
      "          '            0\\n'\n",
      "          '          \\n'\n",
      "          '        \\n'\n",
      "          '      \\n'\n",
      "          '    \\n'\n",
      "          '    {\\\\displaystyle 1/R_{0}}\\n'\n",
      "          '  .\\n'\n",
      "          'The basic reproduction number is affected by several factors, '\n",
      "          'including the duration of infectivity of affected people, the '\n",
      "          'infectiousness of the microorganism, and the number of susceptible '\n",
      "          'people in the population that the infected people contact.'},\n",
      " {'article_title': 'Disease X',\n",
      "  'bert_score': 0.19220882654190063,\n",
      "  'section_title': 'Zoonotic viruses',\n",
      "  'text': 'Disease X,Candidates,Zoonotic viruses\\n'\n",
      "          'On the addition of Disease X in 2018, the WHO said it could come '\n",
      "          'from many sources citing haemorrhagic fevers and the more recent '\n",
      "          'non-polio enterovirus. However, Røttingen speculated that Disease X '\n",
      "          'would be more likely come from zoonotic transmission (an animal '\n",
      "          'virus that jumps to humans), saying: \"It\\'s a natural process and '\n",
      "          'it is vital that we are aware and prepare. It is probably the '\n",
      "          'greatest risk\". WHO special advisor Professor Marion Koopmans, also '\n",
      "          'noted that the rate at which zoonotic diseases were appearing was '\n",
      "          'accelerating, saying: \"The intensity of animal and human contact is '\n",
      "          'becoming much greater as the world develops. This makes it more '\n",
      "          'likely new diseases will emerge but also modern travel and trade '\n",
      "          'make it much more likely they will spread\".'}]\n"
     ]
    }
   ],
   "source": [
    "# using the distilled BERT model\n",
    "reranked = compute_embeddings(query, results, model='distilbert-base-nli-stsb-mean-tokens')\n",
    "pprint(reranked)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "03_Building_SearchAPI_w_Elasticsearch_and_BERT.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
